---
title: "01-methylkit"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Obtain session information

```{r}
sessionInfo()
```

## Download files from `gannet`

```{bash}
#rsync -avz --progress sr320@mox.hyak.uw.edu:/gscratch/srlab/sr320/analyses/2019/0327/*bismark_bt2.deduplicated.sorted.bam ../analyses/
```


```{bash}
curl https://gannet.fish.washington.edu:/seashell/bu-mox/analyses/2019/0327/{*}bismark_bt2.deduplicated.sorted.bam ../analyses/
```


## Check current directory 

```{bash}
pwd
```

## Create list of all bismark data files, which hav reads that are already trimmed and aligned. These BAM files are also sorted and indexed. 

```{r}
file.list_18=list('../data/zr1394_1_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_2_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_3_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',                            		   						 	'../data/zr1394_4_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_5_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',                              '../data/zr1394_6_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_7_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_8_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',                            		   						 	'../data/zr1394_9_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_10_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',                              '../data/zr1394_11_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_12_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_13_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',                            		   						 	'../data/zr1394_14_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_15_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',                              '../data/zr1394_16_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_17_s456_trimmed_bismark_bt2.deduplicated.sorted.bam',
'../data/zr1394_18_s456_trimmed_bismark_bt2.deduplicated.sorted.bam'
)
```

## Require MethylKit 

```{r}
library(methylKit)
```

## The following command reads sorted BAM files and calls methylation percentage per base, and creates a methylRaw object for CpG methylation.
It also assigns minimum coverage of 2x and the treatments (in this case, the Olympia oyster population)

NOTE: according to the below code, an outdated Oly genome version was used in Bismark (most current version is v081). 

```{r, eval = FALSE}
myobj_18 = processBismarkAln(location = file.list_18, sample.id = list("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"), assembly = "v071", read.context="CpG", mincov=2, treatment = c(0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1))
```

## Save the MethylKit object; re-doing the previous step is memory/time intensive, so best to use the saved object moving forward. 

Object is also available at https://d.pr/f/vRtrqZ


```{r, eval = FALSE}
save(myobj_18, file = "../analyses/myobj_18")
```


```{bash}
#zip ../analyses/myobj_18.zip ../analyses/myobj_18
```

???? not sure what this is referencing, if anything 

```{r}
#load("analyses/myobj_52")
```

## Check the basic stats about the methylation data - coverage and percent methylation. Index the object to look through each sample

```{r}
getMethylationStats(myobj_18[[2]],plot=F,both.strands=FALSE)
myobj_18[[5]]
```

```{r}
getMethylationStats(myobj_18[[2]],plot=T,both.strands=FALSE)
```

```{r}
getCoverageStats(myobj_18[[5]],plot=TRUE,both.strands=FALSE)
```

```{r}
getCoverageStats(myobj_18[[13]],plot=TRUE,both.strands=FALSE)
```

## Filter out loci where coverage is less than 10x or greater than 100x. Then, destrand. 

```{r}
filtered.myobj=filterByCoverage(myobj_18,lo.count=10,lo.perc=NULL,
                                      hi.count=100,hi.perc=NULL)
meth_filter=methylKit::unite(filtered.myobj, destrand=TRUE)
save(meth_filter, file = "../analyses/meth_filter") #save object to file 

clusterSamples(meth_filter, dist="correlation", method="ward", plot=TRUE)
PCASamples(meth_filter, screeplot = TRUE)
```


```{r}
myDiff=calculateDiffMeth(meth_filter,mc.cores=4)
```

```{r}
# get hyper methylated bases
myDiff25p.hyper=getMethylDiff(myDiff,difference=25,qvalue=0.01,type="hyper")
#
# get hypo methylated bases
myDiff25p.hypo=getMethylDiff(myDiff,difference=25,qvalue=0.01,type="hypo")
#
#
# get all differentially methylated bases
myDiff25p=getMethylDiff(myDiff,difference=25,qvalue=0.01)
```


```{r}
write.table(myDiff25p, file = "../analyses/myDiff25p.tab", sep = "\t")
```


```{r}
write.table(myDiff50p, file = "../analyses/myDiff50p.tab", sep = "\t")
```



---
# Taking the DMLs to a bed

```{r}
library(readr)
#myDiff25p <- read_csv("../analyses/myDiff25p.csv")

```


```{r}
myDiff25p
```


```{r}
library(tidyverse)
```

Also at http://gannet.fish.washington.edu/seashell/snaps/dml25.bed

```{r}
dml25 <-  mutate(myDiff25p, start = start -1, end = end + 1) %>% select(chr, start, end, meth.diff) %>%
  mutate_if(is.numeric, as.integer)
write_delim(dml25, "../analyses/dml25.bed",  delim = '\t', col_names = FALSE)
```

##  Subset the filtered methylation count dataframe for only those that are differentially methylated. Save object to file. 

```{r}
dml25_counts <- meth_filter %>%
  methylKit::select(meth_filter$chr %in% myDiff25p$chr & 
           meth_filter$start %in% myDiff25p$start & 
           meth_filter$end %in% myDiff25p$end)

class(dml25_counts) #class should be methylBase
save(dml25_counts, file = "../analyses/dml25_counts") #save object to be used in subsequent notebook 
```


# Tiling window analysis 

Summarize methylation information over tiling windows, to then assess differentially methylated regions (DMRs). The function below tiles the genome with windows 1000bp length and 1000bp step-size and summarizes the methylation information on those tiles. In this case, it returns a methylRawList object which can be fed into unite and calculateDiffMeth functions consecutively to get differentially methylated regions. The tilling function adds up C and T counts from each covered cytosine and returns a total C and T count for each tile.

```{r}
#tiles_1000=tileMethylCounts(filtered.myobj,win.size=1000,step.size=1000, mc.cores=4)
head(tiles_1000[[1]],3)

#tiles_200=tileMethylCounts(filtered.myobj,win.size=200,step.size=200, mc.cores=4)
#head(tiles_200)
```

```{r}
meth_tile_filter=methylKit::unite(tiles_1000, destrand=TRUE)
meth_tile_filter 

clusterSamples(meth_tile_filter, dist="correlation", method="ward", plot=TRUE)
PCASamples(meth_tile_filter)
```

## Test for differentially methylated regions 

```{r}
myDiff_tiled=calculateDiffMeth(meth_tile_filter,mc.cores=4)
```

## Test for differentially methylated regions 

```{r}
# get hyper methylated regions 
myDiff_tiled_25p.hyper=getMethylDiff(myDiff_tiled,difference=25,qvalue=0.01,type="hyper")
#
# get hypo methylated regions
myDiff_tiled_25p.hypo=getMethylDiff(myDiff_tiled,difference=25,qvalue=0.01,type="hypo")
#
# get all differentially methylated regions (25% different)
myDiff_tiled_25p=getMethylDiff(myDiff_tiled,difference=25,qvalue=0.01) #only 2 regions! 
```