{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 05-MACAU \n",
    "\n",
    "In this notebook I use MACAU - Mixed model Association for Count data via data AUgmentation - to assess the influence of Olympia oyster size/weight on DNA methylation, while controlling for relatedness. I leverage sequence data from bisulfite treated DNA from 2 Olympia oyster populations - Dabob Bay (in Hood Canal), and Oyster Bay (in South Puget Sound). \n",
    "\n",
    "MACAU runs on Lenox. The software is available on [Zhou Lab website](http://www.xzlab.org/software.html). The R program seems to have some issues, so I'm using the binary version installed by Sam on Roadrunner. The [MACAU user manual](http://www.xzlab.org/software/macau/MACAUmanual.pdf) is the basis for this analysis. \n",
    "\n",
    "### Input files \n",
    "\n",
    "MACAU inputs for this type of analysis include: \n",
    "\n",
    "(1) Methylated read counts (raw counts, not percentages)\n",
    "(2) Total read counts (raw counts)\n",
    "(3) Relatedness matrix\n",
    "(4) Predictor variable, or covariates \n",
    "\n",
    "First, I download all the files I'll need for MACAU. The files were generated as follows: \n",
    "\n",
    "Both read count files - (1) and (2) above - were generated by Laura Spencer in RStudio in the [04-raw-count-files.Rmd notebook](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/code/04-raw-count-files.Rmd), using the MethylKit object that Steven Roberts generated in MethylKit in the [01-methylkit.Rmd notebook](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/code/01-methylkit.Rmd). Note that in both files, counts from + and - strand were combined. These matricese were created from a MethylKit object that had been filtered for all loci with minimum of 10x coverage and maximum of 100x coverage, and with loci that were present in 8 of the 9 samples per population.\n",
    "\n",
    "- [counts-filtered-total.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/counts-filtered-total.txt?raw=true)\n",
    "- [counts-filtered-meth.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/counts-filtered-meth.txt?raw=true)\n",
    "\n",
    "The relatedness matrix - (3) - was generated by Katherine Silliman from 2bRad data, using SNPs and the program ANGSD. Check out her notebook entry, [SOS_angsd.ipynb](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/2bRAD/SOS_angsd.ipynb). \n",
    "- [HSmbdsamples_rab.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/2bRAD/HSmbdsamples_rab.txt) - generated using only HC/SS samples  \n",
    "- [mbdsamples_rab.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/2bRAD/mbdsamples_rab.txt) - using all three populations - HC/SS/NF  \n",
    "\n",
    "The predictor covariates - (4) - is the variable we will use to assess differential DNA methylation. Our data is both shell length and wet weight. \n",
    "- [size.macau.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/data/size.macau.txt)\n",
    "\n",
    "NOTE: all 4 files need to have samples in the same order. They are all ordered using the mbd seq. sample numbers, 1-18. For the count data files, they automatically were ordered sequentially this way when I generated them. For the size covariate file, I re-orded samples in the [04-raw-count-files.Rmd notebook](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/code/04-raw-count-files.Rmd).  Katherine re-ordered her relatedness matrix in the [SOS_angsd.ipynb notebook](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/2bRAD/SOS_angsd.ipynb). \n",
    "\n",
    "Here are screen shots to confirm. \n",
    "\n",
    "_Hold for SSs of other dataframes from RStudio_\n",
    "\n",
    "Katherine's relatedness matrix order \n",
    "<img src=\"attachment:image.png\" width=\"600\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See if we can access MACAU "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      " MACAU version 1.00alpha, released on 06/05/2015\r\n",
      " implemented by Xiang Zhou\r\n",
      "\r\n",
      " type ./macau -h [num] for detailed helps\r\n",
      " options: \r\n",
      " 1: quick guide\r\n",
      " 2: file I/O related\r\n",
      " 3: fit binomial mixed models\r\n",
      " 4: note\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! /home/shared/macau/macau -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " QUICK GUIDE\r\n",
      " To fit binomial mixed models: \r\n",
      "         ./macau -r [filename] -t [filename] -p [filename] -k [filename] -bmm -o [prefix]\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! /home/shared/macau/macau -h 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'/home/srlab/GitHub/paper-oly-mbdbs-gen/code'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# create a macau/ directory  \n",
    "\n",
    "! mkdir ../analyses/macau/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/srlab/GitHub/paper-oly-mbdbs-gen/analyses/macau\n"
     ]
    }
   ],
   "source": [
    "# MACAU writes output files to current directory, so move there \n",
    "%cd ../analyses/macau/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'/home/srlab/GitHub/paper-oly-mbdbs-gen/analyses/macau'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.2\t17.41\r\n",
      "1.9\t20.43\r\n",
      "2.2\t25.33\r\n",
      "1.1\t19.38\r\n",
      "2.2\t26.79\r\n",
      "1.2\t19.8\r\n",
      "2.1\t20.54\r\n",
      "1.9\t19.5\r\n",
      "1.4\t18.43\r\n",
      "2.2\t21.02\r\n"
     ]
    }
   ],
   "source": [
    "# Confirm which column I want to use as predictor variable. \n",
    "#1= covariate file intercept \n",
    "#2 - wet weight (grams, in shell) <-- use this as covariate\n",
    "#3 - shell length (mm) <-- use this as predictor \n",
    "\n",
    "! head ../../data/predictors.size.macau.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the following options: \n",
    "    \n",
    "`-g`  specify the methylated read counts file name.  \n",
    "`-t`  specify the total read counts file name.  \n",
    "`-p`  specify the predictor variable file name.  \n",
    "`-n 2` specify which predictor variable column to use in analysis. In our case column 2 (length in mm)\n",
    "`-k`  specify the kinship/relatedness matrix file name.  \n",
    "`-bmm`  specifies binomial mixed model.  \n",
    "`-o`  specify output file prefix (default “result”).  \n",
    "\n",
    "Run started at 1:15pm on 08/12/2019. Run ended at TBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error! fail to open read count file: ../counts-filtered-meth.txt\r\n",
      "error! fail to open read depth file: ../counts-filtered-total.txt\r\n"
     ]
    }
   ],
   "source": [
    "! /home/shared/macau/macau \\\n",
    "-g ../counts-filtered-meth.txt \\\n",
    "-t ../counts-filtered-total.txt \\\n",
    "-p ../../data/predictors.size.macau.txt \\\n",
    "-n 2 \\\n",
    "-c ../../data/cov.weight.macau.txt \\\n",
    "-k ../2bRAD/HSmbdsamples_rab.txt \\\n",
    "-bmm \\\n",
    "-o 20200205-macau"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;34moutput\u001b[0m/\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id\tn\tacpt_rate\tbeta\tse_beta\tpvalue\th\tse_h\tsigma2\tse_sigma2\talpha0\tse_alpha0\talpha1\tse_alpha1\r\n",
      "Contig038973\t18\t4.069e-01\t-6.949e-02\t5.548e-02\t2.104e-01\t7.713e-01\t2.336e-01\t2.793e+00\t5.452e+00\t1.936e+00\t1.027e+00\t4.199e-01\t2.701e-01\r\n",
      "Contig039226\t18\t4.844e-01\t-1.022e-02\t3.708e-02\t7.829e-01\t6.961e-01\t2.519e-01\t9.479e-01\t1.905e+00\t8.591e-01\t6.584e-01\t-2.180e-01\t1.730e-01\r\n",
      "Contig039234\t18\t3.831e-01\t4.603e-02\t3.661e-02\t2.087e-01\t7.543e-01\t2.308e-01\t3.532e+00\t1.117e+01\t-1.400e+00\t7.399e-01\t-1.432e-01\t1.573e-01\r\n",
      "Contig039252\t18\t4.388e-01\t-4.211e-02\t5.445e-02\t4.393e-01\t7.156e-01\t2.512e-01\t3.239e+00\t8.253e+00\t1.283e+00\t1.019e+00\t1.868e-01\t2.675e-01\r\n",
      "Contig041234\t18\t3.909e-01\t7.483e-02\t9.470e-02\t4.294e-01\t6.285e-01\t2.689e-01\t5.002e+00\t8.564e+00\t4.363e-01\t1.708e+00\t-4.537e-01\t3.745e-01\r\n",
      "Contig064124\t18\t3.505e-01\t5.754e-02\t9.419e-02\t5.413e-01\t4.358e-01\t2.380e-01\t4.554e+00\t6.052e+00\t-3.804e-01\t1.737e+00\t-4.357e-01\t4.309e-01\r\n",
      "Contig064179\t18\t2.235e-01\t1.405e-01\t7.007e-02\t4.495e-02\t8.094e-01\t1.432e-01\t1.559e+01\t1.805e+01\t4.177e-01\t1.547e+00\t-6.862e-01\t2.475e-01\r\n",
      "Contig064210\t18\t2.332e-01\t8.226e-03\t1.060e-01\t9.381e-01\t7.820e-01\t1.713e-01\t1.350e+01\t1.682e+01\t2.864e+00\t1.900e+00\t1.481e-02\t5.874e-01\r\n",
      "Contig064217\t18\t4.165e-01\t-3.517e-02\t5.733e-02\t5.395e-01\t7.354e-01\t2.509e-01\t2.946e+00\t5.370e+00\t2.799e+00\t1.072e+00\t-3.354e-02\t2.646e-01\r\n"
     ]
    }
   ],
   "source": [
    "# Peak at the results file \n",
    "! head output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id\tn\tacpt_rate\tbeta\tse_beta\tpvalue\th\tse_h\tsigma2\tse_sigma2\talpha0\tse_alpha0\talpha1\tse_alpha1\r\n"
     ]
    }
   ],
   "source": [
    "# List column names - p-value is 6th column (use for indexing)\n",
    "! head -n 1 output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15616\r\n"
     ]
    }
   ],
   "source": [
    "# Count # hits with p-value <0.05\n",
    "! awk '$6<0.05{pvalue++} END{print pvalue+0}' \\\n",
    "output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4226\r\n"
     ]
    }
   ],
   "source": [
    "# Count # hits with p-value <0.01\n",
    "! awk '$6<0.01{pvalue++} END{print pvalue+0}' \\\n",
    "output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "849\r\n"
     ]
    }
   ],
   "source": [
    "# Count # hits with p-value <0.001\n",
    "! awk '$6<0.001{pvalue++} END{print pvalue+0}' \\\n",
    "output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##\r\n",
      "## MACAU Version = 1.00alpha\r\n",
      "##\r\n",
      "## Command Line Input = /home/shared/macau/macau -g ../counts-meth-destrand.txt -t ../counts-total-destrand.txt -p ../../data/predictors.size.macau.txt -n 2 -c ../../data/cov.weight.macau.txt -k ../2bRAD/HSmbdsamples_rab.txt -bmm -o 20190812-macau \r\n",
      "##\r\n",
      "## Date = Mon Aug 12 18:44:58 2019\r\n",
      "##\r\n",
      "## Summary Statistics:\r\n",
      "## number of total individuals = 18\r\n",
      "## number of analyzed individuals = 18\r\n",
      "## number of total genes/sites = 256043\r\n",
      "## number of analyzed genes/sites = 512086\r\n",
      "## number of covariates = 2\r\n",
      "##\r\n",
      "## Computation time:\r\n",
      "## total computation time = 331.402 min \r\n",
      "## computation time break down: \r\n",
      "##      time on eigen-decomposition = 7.16685 min \r\n",
      "##      time on calculating matrix-vector multiplication = 38.4124 min \r\n",
      "##      time on sampling Z = 21.0581 min \r\n",
      "##      time on sampling MW = 32.2596 min \r\n",
      "##      time on sampling Hyperparameters = 218.413 min \r\n",
      "##      time on sampling BAU = 30.8721 min \r\n",
      "##\r\n"
     ]
    }
   ],
   "source": [
    "# Review run log \n",
    "! cat output/20200205-macau.log.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
