---
title: "Locations of DMLs and MACAU loci"
author: "Laura H Spencer"
date: "11/20/2019"
output: html_document
---

### Load libraries 

```{r setup, message=FALSE, warning=FALSE, results=FALSE}
list.of.packages <- c("tidyverse", "reshape2", "here") #add new libraries here 
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})
sessionInfo()
```

## 1. DMLs 

### Where are differentially methylated loci located? 

```{r, message=FALSE, warning=FALSE}
DMLfiles <- c("DML-CDS.bed", "DML-exon.bed", "DML-gene.bed", "DML-mRNA.bed", "DML-TE.bed", "DML-ASV.bed", "DML-intragenic.bed")
DML.features <- list()
for (i in c(1:7)) {
  DML.features[[i]] <- read_delim(here::here("analyses", "BEDtools", DMLfiles[i]), delim = '\t', col_names = FALSE) %>% as_tibble()}
for (i in 1:6) {
  DML.features[[i]] <- DML.features[[i]] %>%
    setNames(c("contig.dml","start.dml","end.dml","score.dml","contig.feat", "source.feat","feature","start.feat","end.feat","unknown1","strand","unknown2","attribute")) %>%
mutate(ID=str_extract(attribute, "ID=(.*?);"),
       Parent=str_extract(attribute, "Parent=(.*?);"),
       Name=str_extract(attribute, "Name=(.*?);"),
       Alias=str_extract(attribute, "Alias=(.*?);"),
       AED=str_extract(attribute, "AED=(.*?);"),
       eAED=str_extract(attribute, "eAED=(.*?);"),
       Note=str_extract(attribute, "Note=(.*?);"),
       Ontology_term=str_extract(attribute, "Ontology_term=(.*?);"),
       Dbxref=str_extract(attribute, "Dbxref=(.*?);")
       ) %>%
mutate_at("feature", as.factor)
}
names(DML.features) <- c("DML.CDS", "DML.exon", "DML.gene", "DML.mRNA", "DML.TE", "DML.ASV", "DML.intragenic")
DML.features[["DML.intragenic"]] <- 
  DML.features[["DML.intragenic"]] %>% 
  setNames(c("contig.DML", "start.dml", "end.dml", "score.dml"))

DML.features[[1]] <- DML.features[[1]] %>% mutate_at("unknown2", as.character)
DML.features[[5]] <- DML.features[[5]] %>% mutate_at("unknown1", as.character)
DML.features[[6]] <- DML.features[[6]] %>% mutate_at("unknown1", as.character)
DML.features.df <- bind_rows(DML.features[-6:-7])
print(DML.summary <- table(DML.features.df[c("feature")])) #Note: "similarity" refers to transposable elements; also NO alternative splice variants are included. 

DML.features.df
```

## 2. MACAU loci

### Where are differentially methylated loci located? 

```{r, message=FALSE, warning=FALSE}
macau.files <- c("macau-CDS.bed", "macau-exon.bed", "macau-gene.bed", "macau-mRNA.bed", "macau-ASV.bed", "macau-intragenic.bed")
macau.features <- list()
for (i in c(1:6)) {
  macau.features[[i]] <- read_delim(here::here("analyses", "BEDtools", macau.files[i]), delim = '\t', col_names = FALSE) %>% as_tibble()}
for (i in 1:5) {
  macau.features[[i]] <- macau.features[[i]] %>%
    setNames(c("contig.macau","start.macau","end.macau","contig.feat", "source.feat","feature","start.feat","end.feat","unknown1","strand","unknown2","attribute")) %>%
mutate(ID=str_extract(attribute, "ID=(.*?);"),
       Parent=str_extract(attribute, "Parent=(.*?);"),
       Name=str_extract(attribute, "Name=(.*?);"),
       Alias=str_extract(attribute, "Alias=(.*?);"),
       AED=str_extract(attribute, "AED=(.*?);"),
       eAED=str_extract(attribute, "eAED=(.*?);"),
       Note=str_extract(attribute, "Note=(.*?);"),
       Ontology_term=str_extract(attribute, "Ontology_term=(.*?);"),
       Dbxref=str_extract(attribute, "Dbxref=(.*?);")
       ) %>%
mutate_at("feature", as.factor)
}
names(macau.features) <- c("macau.CDS", "macau.exon","macau.gene", "macau.mRNA", "macau.ASV", "macau.intragenic")
macau.features[["macau.intragenic"]] <- 
  macau.features[["macau.intragenic"]] %>% 
  setNames(c("contig.macau", "start.macau", "end.macau"))

macau.features[[1]] <- macau.features[[1]] %>% mutate_at("unknown2", as.character)
macau.features[[5]] <- macau.features[[5]] %>% mutate_at("unknown1", as.character)
macau.features.df <- bind_rows(macau.features[-5:-6])
print(macau.summary <- table(macau.features.df[c("feature")])) #NOTE this does not contain alternative splice variants 

macau.features[[3]]
```

Are any MACAU associated with DMLs / DMGs ? 

```{r}
# Barplots of all MACAU loci  

meth_filter_calcs <- meth_filter_reshaped %>% 
  group_by(population, chr, start) %>% 
  dplyr::summarise(
    mean_percMeth = mean(percMeth, na.rm=TRUE),
    sd_percMeth=sd(percMeth, na.rm=TRUE),
    n()) 

meth_filter_calcs %>% 
      filter(chr %in% macau.features[[3]]$contig.macau &
           start %in% macau.features[[3]]$start.macau) %>%
  ggplot(aes(x = population, y = mean_percMeth, fill = population, 
                         label=paste0(round(mean_percMeth, digits = 2), "%"))) + 
      geom_bar(stat="identity", width = 0.5) + ylim(0,110) +
      geom_pointrange(aes(ymin=mean_percMeth, 
                        ymax=mean_percMeth+sd_percMeth, width=0.15)) + 
      geom_text(size=3, vjust=-0.5, hjust=1.25) +
      theme_light() + ggtitle("") + facet_wrap(~chr) + 
    scale_fill_manual(values=c("firebrick3","dodgerblue3"))

```



```{r}
macau.features.df
```


## 3. All loci 

### Where are ALL methylated loci located? 

```{r, message=FALSE, warning=FALSE}
allLocifiles <- c("AllLoci-CDS.bed", "AllLoci-exon.bed", "AllLoci-gene.bed", "AllLoci-mRNA.bed", "AllLoci-TE.bed", "AllLoci-ASV.bed", "AllLoci-intragenic.bed")
allLoci.features <- list()
for (i in c(1:7)) {
  allLoci.features[[i]] <- read_delim(here::here("analyses","BEDtools", allLocifiles[i]), delim = '\t', col_names = FALSE) %>% as_tibble()}
for (i in 1:6) {
  allLoci.features[[i]] <- allLoci.features[[i]] %>%
    setNames(c("contig.allLoci","start.allLoci","end.allLoci", "contig.feat", "source.feat","feature","start.feat","end.feat","unknown1","strand","unknown2","attribute")) %>%
mutate(ID=str_extract(attribute, "ID=(.*?);"),
       Parent=str_extract(attribute, "Parent=(.*?);"),
       Name=str_extract(attribute, "Name=(.*?);"),
       Alias=str_extract(attribute, "Alias=(.*?);"),
       AED=str_extract(attribute, "AED=(.*?);"),
       eAED=str_extract(attribute, "eAED=(.*?);"),
       Note=str_extract(attribute, "Note=(.*?);"),
       Ontology_term=str_extract(attribute, "Ontology_term=(.*?);"),
       Dbxref=str_extract(attribute, "Dbxref=(.*?);")
       ) %>%
mutate_at("feature", as.factor)
}
names(allLoci.features) <- c("allLoci.CDS","allLoci.exon","allLoci.gene","allLoci.mRNA","allLoci.TE","allLoci.ASV","allLoci.intragenic")
allLoci.features[["allLoci.intragenic"]] <- 
  allLoci.features[["allLoci.intragenic"]] %>% 
  setNames(c("contig.allLoci", "start.allLoci", "end.allLoci"))
allLoci.features[[1]] <- allLoci.features[[1]] %>% mutate_at("unknown2", as.character)
allLoci.features[[6]] <- allLoci.features[[6]] %>% mutate_at("unknown1", as.character)
allLoci.features.df <- bind_rows(allLoci.features[1:5]) #don't include alternative splice variants or intragenic 
print(allLoci.summary <- table(allLoci.features.df[c("feature")])) #Note, "similartiy"= transposable elements 
```

## Summary 

```{r, message=FALSE, warning=FALSE}
summary <- merge(x=merge(x=melt(DML.summary, varnames = "feature", value.name = "DML"), 
      y=melt(macau.summary, varnames = "feature", value.name = "macau"),
      by="feature", all=TRUE), 
      y=melt(allLoci.summary, varnames = "feature", value.name = "allLoci"),
      by="feature", all=TRUE)
summaryL <- melt(summary, variable.name = "analysis", value.name = "count")
summaryL$feature <- as.factor(str_replace(summaryL$feature, "similarity", "TE"))
summaryL <- summaryL %>% 
  na.omit() %>%
  group_by(analysis) %>%
  mutate(percent=100*(count/sum(count)))

ggplot(data=subset(summaryL, analysis!="macau"), aes(x=feature, y=log(count), fill=analysis)) +
  geom_bar(stat="identity") + 
  scale_fill_brewer(palette="Paired") + ggtitle("DMLs loci and all methylated loci, Log-counts") +
  theme_minimal()

ggplot(data=subset(summaryL, analysis!="DML"), aes(x=feature, y=log(count), fill=analysis)) +
  geom_bar(stat="identity") +
  scale_fill_brewer(palette="Paired") + ggtitle("MACAU and all methylated loci, Log-counts") +
  theme_minimal()

ggplot(data=subset(summaryL, analysis!="macau"), aes(x=feature, y=percent, fill=analysis)) +
  geom_bar(stat="identity", position=position_dodge()) + 
  scale_fill_brewer(palette="Paired") + ggtitle("% of DML loci, and all loci, in each feature") +
  geom_text(aes(label=paste(round(percent), "%", sep="")), vjust=-0.5, color="gray30",
            position = position_dodge(0.9), size=3)+
  theme_minimal()

ggplot(data=subset(summaryL, analysis!="DML"), aes(x=feature, y=percent, fill=analysis)) +
  geom_bar(stat="identity", position=position_dodge()) +
  scale_fill_brewer(palette="Paired") + ggtitle("% of MACAU loci, and all loci, in each feature") +
  geom_text(aes(label=paste(round(percent), "%", sep="")), vjust=-0.5, color="gray30",
            position = position_dodge(0.9), size=3) +
  theme_minimal()
```

## Extract GO terms 

```{r}
macau.GO <- macau.features.df[c("contig.macau", "start.macau", "end.macau", "feature", "start.feat", "end.feat", "Note", "Ontology_term")] %>%
  distinct(contig.macau, start.macau, Note, Ontology_term, .keep_all = TRUE) %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern="Ontology_term=",replacement = "")) %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern=";",replacement = "")) %>%
  separate(Ontology_term, sep=",", into=paste("GO", 1:6, sep="_")) %>%
  pivot_longer(cols=c("GO_1","GO_2","GO_3","GO_4","GO_5","GO_6"), names_to = "GO_number", values_to = "GO_term") %>%
  select(-GO_number) %>%
  filter(!is.na(Note) & !is.na(GO_term))
write_csv(macau.features.df, path = here::here("analyses/", "macau/", "macau.features.csv")) #write out df with all MACAU features 
write(macau.GO$GO_term, file = here::here("analyses/", "macau/", "macau.GO.txt")) #write out df with just GO terms 
write.table(subset(macau.features.df, feature=="gene"), here::here("analyses/", "macau/", "MACAU.geneinfo.txt"), sep = "\t") #write out gene info for MACAU loci to associate with UniprotIDs

DML.GO <- DML.features.df[c("contig.dml", "start.dml", "end.dml", "feature", "start.feat", "end.feat", "Note", "Ontology_term")] %>%
  distinct(contig.dml, start.dml, Note, Ontology_term, .keep_all = TRUE) %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern="Ontology_term=",replacement = "")) %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern=";",replacement = "")) %>%
  separate(Ontology_term, sep=",", into=paste("GO", 1:6, sep="_")) %>%
  pivot_longer(cols=c("GO_1","GO_2","GO_3","GO_4","GO_5","GO_6"), names_to = "GO_number", values_to = "GO_term") %>%
  select(-GO_number) %>%
  filter(!is.na(Note) & !is.na(GO_term)) 
write_csv(DML.features.df, path = here::here("analyses/", "DML.features.txt")) #write out df with all DML features 
write(DML.GO$GO_term, file = here::here("analyses/", "DML.GO.txt"))
write.table(subset(DML.features.df, feature=="gene"), here::here("analyses/", "DML.geneinfo.txt"), sep = "\t") #write out gene info for DMLs to associate with UniprotIDs

allLoci.GO <- allLoci.features.df[c("contig.allLoci", "start.allLoci", "end.allLoci", "feature", "start.feat", "end.feat", "Note", "Ontology_term")] %>%
  distinct(contig.allLoci, start.allLoci, Note, Ontology_term, .keep_all = TRUE) %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern="Ontology_term=",replacement = "")) %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern=";",replacement = "")) %>%
  separate(Ontology_term, sep=",", into=paste("GO", 1:6, sep="_")) %>%
  pivot_longer(cols=c("GO_1","GO_2","GO_3","GO_4","GO_5","GO_6"), names_to = "GO_number", values_to = "GO_term") %>%
  select(-GO_number) %>%
  filter(!is.na(Note) & !is.na(GO_term)) 
write_csv(allLoci.features.df, path = here::here("analyses/", "AllLoci.features.txt")) #write out df with all loci features 
write(allLoci.GO$GO_term, file = here::here("analyses/", "allLoci.GO.txt"))
write.table(subset(allLoci.features.df, feature=="gene"), here::here("analyses/", "macau/", "allLoci.geneinfo.txt"), sep = "\t") #write out gene info for All loci fed into MACAU to associate with UniprotIDs

```

## Create separate dataframes with annotation information for all hypomethylated loci in Hood Canal & South Sound, separately 

```{r}
hypo_HC_genes <-
  DML.features.df %>% 
  filter(contig.dml %in% hypo_HC$chr &
         start.dml %in% 1+hypo_HC$start & 
         feature == "gene" & 
         Note != "Note=Protein of unknown function;") %>% 
  distinct(contig.dml, start.dml, Note, Ontology_term, .keep_all = TRUE)
  
hypo_HC_GO <- hypo_HC_genes %>%
mutate(Ontology_term = str_replace(Ontology_term, pattern="Ontology_term=",replacement = "")) %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern=";",replacement = "")) %>% 
  separate(Ontology_term, sep=",", into=paste("GO", 1:3, sep="_")) %>%
  pivot_longer(cols=c("GO_1","GO_2","GO_3"), names_to = "GO_number", values_to = "GO_term") %>%
  select(-GO_number) %>%
  filter(!is.na(Note) & !is.na(GO_term)) 
write(hypo_HC_GO$GO_term, file = here::here("analyses/", "hypo_HC_GO.txt"))


hypo_SS_genes <- DML.features.df %>% 
  filter(contig.dml %in% hypo_SS$chr &
         start.dml %in% 1+hypo_SS$start & 
           feature == "gene" & 
           Note != "Note=Protein of unknown function;") %>% 
  distinct(contig.dml, start.dml, Note, Ontology_term, .keep_all = TRUE)

hypo_SS_GO <- hypo_SS_genes %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern="Ontology_term=",replacement = "")) %>%
  mutate(Ontology_term = str_replace(Ontology_term, pattern=";",replacement = "")) %>% 
  separate(Ontology_term, sep=",", into=paste("GO", 1:5, sep="_")) %>%
  pivot_longer(cols=c("GO_1","GO_2","GO_3", "GO_4", "GO_5"), names_to = "GO_number", values_to = "GO_term") %>%
  select(-GO_number) %>%
  filter(!is.na(Note) & !is.na(GO_term)) 
write(hypo_HC_GO$GO_term, file = here::here("analyses/", "hypo_SS_GO.txt"))


# Barplots of differentially methylated loci that are located in known genes 

hypo_HC %>%
   filter(chr %in% hypo_HC_genes$contig.dml &
         start %in% hypo_HC_genes$start.dml-1) %>%
ggplot(aes(x = population, y = mean_percMeth, fill = population, 
                         label=paste0(round(mean_percMeth, digits = 2), "%"))) + 
      geom_bar(stat="identity", width = 0.5) + ylim(0,110) +
      geom_pointrange(aes(ymin=mean_percMeth, 
                        ymax=mean_percMeth+sd_percMeth, width=0.15)) + 
      geom_text(size=3, vjust=-0.5, hjust=1.1) +
      theme_minimal() + xlab("Population") + ylab("% Methylation") +
  scale_y_continuous(breaks=c(0,25,50,75,100)) +
    scale_fill_manual(values=c("firebrick3","dodgerblue3")) + 
   facet_wrap_paginate(~chr + start, nrow=2, ncol=3, page=2)


hypo_SS %>%
   filter(chr %in% hypo_SS_genes$contig.dml &
         start %in% hypo_SS_genes$start.dml-1) %>%
ggplot(aes(x = population, y = mean_percMeth, fill = population, 
                         label=paste0(round(mean_percMeth, digits = 2), "%"))) + 
      geom_bar(stat="identity", width = 0.5) + ylim(0,110) +
      geom_pointrange(aes(ymin=mean_percMeth, 
                        ymax=mean_percMeth+sd_percMeth, width=0.15)) + 
      geom_text(size=3, vjust=-0.5, hjust=1.1) +
      theme_minimal() + xlab("Population") + ylab("% Methylation") +
  scale_y_continuous(breaks=c(0,25,50,75,100)) +
    scale_fill_manual(values=c("firebrick3","dodgerblue3")) + 
   facet_wrap_paginate(~chr + start, nrow=2, ncol=3, page=2)
```
