---
title: "11-Meth-Gen-Integration"
author: "Laura H Spencer"
date: "4/23/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load libraries

```{r message=FALSE, warning=FALSE, results=FALSE}
list.of.packages <- c("tidyverse", "reshape2", "here", "Pstat") #add new libraries here 

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})
```

```{r}
#analyses/2bRAD/HCSS_Afilt32m70_01_pp90_m75_BSouts.vcf
```

# Calculate methylation Pst for genome bins 

## 1kb Genome Bins 

### Load filtered methylation data object with sample info which was created in the notebook "01-methylkit" 

```{r}
load(here::here("analyses", "methylation-filtered", "R-objects", "meth_filter_reshaped")) 
head(meth_filter_reshaped) #note, this object was used to create the "meth_filter_long.tab" file, so same content  
```

Note: the file Katherine sent didn't have columns separated by tabs, but by single-spaces. I tried without succes to convert spaces to tabs - the closest I got was the below script, but `sed` wouldn't recognize `\t` as a tab. So, I just opened the file in the program Text Edit, did find/replace, and saved as a new file to the genome-features folder in the repo. 

`sed -e 's/ /\t/g' "../analyses/2bRAD/HCSS_sfsm70_10kb.bed" > "../genome-features/Olurida_v081_bins-10kb.bed"`

### Use intersectBed to find where methylated loci and 1kb genome bins (that contain genetic data) intersect
  
  wb: Print all lines in the second file
  a: input data file, which was created in previous code chunk
  b: save annotated gene list

```{bash}
intersectBed \
  -wb \
  -a "../analyses/methylation-filtered/meth_filter_long.tab" \
  -b "../genome-features/Olurida_v081_bins-1kb.bed" \
  > "../analyses/methylation-filtered/meth_filter_1kb-bin.tab"
```

### Count various things 

```{bash}
wc -l "../analyses/methylation-filtered/meth_filter_long.tab"
wc -l "../analyses/methylation-filtered/meth_filter_1kb-bin.tab" 
wc -l "../genome-features/Olurida_v081_bins-1kb.bed"
```

### Check out format of resulting gene regions 

```{bash}
head -n 5  "../analyses/methylation-filtered/meth_filter_1kb-bin.tab" 
```


```{r}
meth_filter_bins_1kb <- read_delim(file = "../analyses/methylation-filtered/meth_filter_1kb-bin.tab", delim = "\t", col_names = c(colnames(meth_filter_reshaped), "contig", "bin.start", "bin.end")) %>% 
  mutate(bin=paste(contig, bin.start, bin.end, sep="_"))
```


### Calculate Pst for all bins represented in our methylation data 

#### First, how many 1kb bins do we have methylation data for? 

```{r}
meth_filter_bins_1kb %>% 
       distinct(bin) %>%
  nrow()
```

#### I need the % methylation data for each of the gene regions, so I'll first create that dataframe. 

```{r}
perc_meth_bins_1kb <- meth_filter_bins_1kb %>% 
   group_by(population, sample, bin, contig, bin.start, bin.end) %>%
   dplyr::summarise(
    mean_percMeth = mean(percMeth, na.rm=TRUE),
    sd_percMeth=sd(percMeth, na.rm=TRUE),
    n())  

# check a couple loci to make sure % methylation is calculated separately for each sample and gene region 
perc_meth_bins_1kb %>% filter(bin=="Contig10074_5000_6000") %>%
  ggplot(aes(x=sample, y=mean_percMeth)) + geom_bar(stat="identity")

perc_meth_bins_1kb %>% filter(bin=="Contig45217_3000_4000") %>%
  ggplot(aes(x=sample, y=mean_percMeth)) + geom_bar(stat="identity")

# Reshape data. Need to have one row per sample, one column with the population, and separate columns with each gene region with % methylation. 
perc_meth_bins_1kb_wide <- perc_meth_bins_1kb %>% 
  ungroup() %>%
  #tidyr::unite("gene_region", c("contig_gene", "start_gene_2kb", "end_gene_2kb"), sep = "_", remove = FALSE) %>%
  dplyr::select(population, sample, bin, mean_percMeth) %>%
  spread(bin, mean_percMeth) %>%
  tibble::column_to_rownames(var = "sample")

head(perc_meth_bins_1kb_wide[1:4]) #confirm correct format 
ncol(perc_meth_bins_1kb_wide)-1 # 71 bins
perc_meth_bins_1kb$`n()` %>% hist()
perc_meth_bins_1kb$`n()` %>% summary()
```

### Run Pst 

```{r}
#Now run the following line and it will provide Pst estimates for every gene.
perc_meth_bins_1kb_Pst <- Pst(perc_meth_bins_1kb_wide, ci=1) #ci=1 to include 95% confidence interval calc 

# Save 1kb Pst object to file for 
save(perc_meth_bins_1kb_Pst, file="../analyses/methylation-filtered/R-objects/perc_meth_bins_1kb_Pst")

# Check out Pst distribution 
hist(perc_meth_bins_1kb_Pst$Pst_Values)
summary(perc_meth_bins_1kb_Pst$Pst_Values)
mean(perc_meth_bins_1kb_Pst$Pst_Values)
sd(perc_meth_bins_1kb_Pst$Pst_Values)
nrow(perc_meth_bins_1kb_Pst)

# format of dataframe that I will save 
perc_meth_bins_1kb_Pst %>% separate(Quant_Varia, into=c("contig","start.bin", "end.bin"), sep = "_", remove = FALSE) %>% noquote() %>% View()

# Write out Pst results 
write.table(perc_meth_bins_1kb_Pst %>% 
              separate(Quant_Varia, into=c("contig","start.bin", "end.bin"), sep = "_", remove = FALSE), 
            file = here::here("analyses", "methylation-filtered", "Pst_bins_1kb.tab"), 
            sep = '\t', na = "NA", row.names = FALSE, col.names = TRUE, quote=FALSE)
```

## 10kb bins 

```{bash}
intersectBed \
  -wb \
  -a "../analyses/methylation-filtered/meth_filter_long.tab" \
  -b "../genome-features/Olurida_v081_bins-10kb.bed" \
  > "../analyses/methylation-filtered/meth_filter_10kb-bin.tab"
```

```{r}
meth_filter_bins_10kb <- read_delim(file = "../analyses/methylation-filtered/meth_filter_10kb-bin.tab", delim = "\t", col_names = c(colnames(meth_filter_reshaped), "contig", "bin.start", "bin.end")) %>% 
  mutate(bin=paste(contig, bin.start, bin.end, sep="_"))

meth_filter_bins_10kb %>% 
       distinct(bin) %>%
  nrow() #271 10kb bins 

perc_meth_bins_10kb <- meth_filter_bins_10kb %>% 
   group_by(population, sample, bin, contig, bin.start, bin.end) %>%
   dplyr::summarise(
    mean_percMeth = mean(percMeth, na.rm=TRUE),
    sd_percMeth=sd(percMeth, na.rm=TRUE),
    n())  

# check a couple loci to make sure % methylation is calculated separately for each sample and gene region 
perc_meth_bins_10kb %>% filter(bin=="Contig10074_0_6621") %>%
  ggplot(aes(x=sample, y=mean_percMeth)) + geom_bar(stat="identity")

perc_meth_bins_10kb %>% filter(bin=="Contig19869_0_10000") %>%
  ggplot(aes(x=sample, y=mean_percMeth)) + geom_bar(stat="identity")

# Reshape data. Need to have one row per sample, one column with the population, and separate columns with each gene region with % methylation. 
perc_meth_bins_10kb_wide <- perc_meth_bins_10kb %>% 
  ungroup() %>%
  #tidyr::unite("gene_region", c("contig_gene", "start_gene_2kb", "end_gene_2kb"), sep = "_", remove = FALSE) %>%
  dplyr::select(population, sample, bin, mean_percMeth) %>%
  spread(bin, mean_percMeth) %>%
  tibble::column_to_rownames(var = "sample")

#Now run the following line and it will provide Pst estimates for every gene.
perc_meth_bins_10kb_Pst <- Pst(perc_meth_bins_10kb_wide, ci=1) #ci=1 to include 95% confidence interval calc 

# Check out Pst distribution 
hist(perc_meth_bins_10kb_Pst$Pst_Values)
summary(perc_meth_bins_10kb_Pst$Pst_Values)
mean(perc_meth_bins_10kb_Pst$Pst_Values)
sd(perc_meth_bins_10kb_Pst$Pst_Values)

# Write out Pst results 
write.table(
  perc_meth_bins_10kb_Pst %>% 
    separate(Quant_Varia, into=c("contig","start.bin", "end.bin"), sep = "_", remove = FALSE), 
  file = here::here("analyses", "methylation-filtered", "Pst_bins_10kb.tab"), 
  sep = '\t', na = "NA", row.names = FALSE, col.names = TRUE, quote=FALSE)
```

## Now I want to calculate Pst for ALL bins, not just those that also contain genetic data 

### Generate .bed files with 1kb and 10kb bins 
Use previously created genome fai file as the bedfile input

```{bash}
# 1kb bins 
awk 'FS=OFS="\t"{print $1, 0, $2}' ../genome-features/Olurida_v081.fa.fai \
| bedtools makewindows -b - -w 1000 > ../genome-features/Olurida_v081_1kbIntervals.bed 

# 10kb bins 
awk 'FS=OFS="\t"{print $1, 0, $2}' ../genome-features/Olurida_v081.fa.fai \
| bedtools makewindows -b - -w 10000 > ../genome-features/Olurida_v081_10kbIntervals.bed 
```

## Assess 1kb bins 

### Use intersectBed to find where methylated loci and all 1kb genome bins intersect
```{bash}
intersectBed \
  -wb \
  -a "../analyses/methylation-filtered/meth_filter_long.tab" \
  -b "../genome-features/Olurida_v081_1kbIntervals.bed" \
  > "../analyses/methylation-filtered/meth_filter_1kb-bin-allmeth.tab"
```

### Import 1kb bin information for all methylation data (not just those that are in bins w/ genetic data)  

```{r}
meth_filter_bins_allmeth_1kb <- read_delim(file = "../analyses/methylation-filtered/meth_filter_1kb-bin-allmeth.tab", delim = "\t", col_names = c(colnames(meth_filter_reshaped), "contig", "bin.start", "bin.end")) %>% 
  mutate(bin=paste(contig, bin.start, bin.end, sep="_"))
```


### Calculate Pst for all bins represented in our methylation data 

#### First, how many 1kb bins do we have methylation data for? 

```{r}
meth_filter_bins_allmeth_1kb %>% 
       distinct(bin) %>%
  nrow()
```

#### I need the % methylation data for each of the gene regions, so I'll first create that dataframe. 

```{r}
perc_meth_bins_allmeth_1kb <- meth_filter_bins_allmeth_1kb %>% 
   group_by(population, sample, bin, contig, bin.start, bin.end) %>%
   dplyr::summarise(
    mean_percMeth = mean(percMeth, na.rm=TRUE),
    sd_percMeth=sd(percMeth, na.rm=TRUE),
    n())  

# Reshape data. Need to have one row per sample, one column with the population, and separate columns with each gene region with % methylation. 
perc_meth_bins_allmeth_1kb_wide <- perc_meth_bins_allmeth_1kb %>% 
  ungroup() %>%
  #tidyr::unite("gene_region", c("contig_gene", "start_gene_2kb", "end_gene_2kb"), sep = "_", remove = FALSE) %>%
  dplyr::select(population, sample, bin, mean_percMeth) %>%
  spread(bin, mean_percMeth) %>%
  tibble::column_to_rownames(var = "sample")
```

### Run Pst for 1kb bins (all with methylation data, not just bins with genetic+meth data)

```{r}
#Now run the following line and it will provide Pst estimates for every gene.
perc_meth_bins_allmeth_1kb_Pst <- Pst(perc_meth_bins_allmeth_1kb_wide) 

# Save 1kb Pst object to file for 
save(perc_meth_bins_allmeth_1kb_Pst, file="../analyses/methylation-filtered/R-objects/perc_meth_bins_allmeth_1kb_Pst")

# Check out Pst distribution 
hist(perc_meth_bins_allmeth_1kb_Pst$Pst_Values)
summary(perc_meth_bins_allmeth_1kb_Pst$Pst_Values)
mean(perc_meth_bins_allmeth_1kb_Pst$Pst_Values)
sd(perc_meth_bins_allmeth_1kb_Pst$Pst_Values)
nrow(perc_meth_bins_allmeth_1kb_Pst)

# format of dataframe that I will save 
perc_meth_bins_allmeth_1kb_Pst %>% separate(Quant_Varia, into=c("contig","start.bin", "end.bin"), sep = "_", remove = FALSE) %>% noquote()

# Write out Pst results 
# NOTE: this file ("Pst_bins_allmeth_1kb.tab") differs from the "Pst_bins_1kb.tab" file in that it contains Pst values for ALL 1kb bins, not just bins that also contain genetic data (many more bins!)
write.table(perc_meth_bins_allmeth_1kb_Pst %>% 
              separate(Quant_Varia, into=c("contig","start.bin", "end.bin"), sep = "_", remove = FALSE), 
            file = here::here("analyses", "methylation-filtered", "Pst_bins_allmeth_1kb.tab"), 
            sep = '\t', na = "NA", row.names = FALSE, col.names = TRUE, quote=FALSE)
```

## Assess 10kb bins 

### Use intersectBed to find where methylated loci and all 1kb genome bins intersect
```{bash}
intersectBed \
  -wb \
  -a "../analyses/methylation-filtered/meth_filter_long.tab" \
  -b "../genome-features/Olurida_v081_10kbIntervals.bed" \
  > "../analyses/methylation-filtered/meth_filter_10kb-bin-allmeth.tab"
```

### Import 10kb bin information for all methylation data (not just those that are in bins w/ genetic data)  

```{r}
meth_filter_bins_allmeth_10kb <- read_delim(file = "../analyses/methylation-filtered/meth_filter_10kb-bin-allmeth.tab", delim = "\t", col_names = c(colnames(meth_filter_reshaped), "contig", "bin.start", "bin.end")) %>% 
  mutate(bin=paste(contig, bin.start, bin.end, sep="_"))
```


### Calculate Pst for all bins represented in our methylation data 

#### First, how many 10kb bins do we have methylation data for? 

```{r}
meth_filter_bins_allmeth_10kb %>% 
       distinct(bin) %>%
  nrow()
```

#### I need the % methylation data for each of the gene regions, so I'll first create that dataframe. 

```{r}
perc_meth_bins_allmeth_10kb <- meth_filter_bins_allmeth_10kb %>% 
   group_by(population, sample, bin, contig, bin.start, bin.end) %>%
   dplyr::summarise(
    mean_percMeth = mean(percMeth, na.rm=TRUE),
    sd_percMeth=sd(percMeth, na.rm=TRUE),
    n())  

# Reshape data. Need to have one row per sample, one column with the population, and separate columns with each gene region with % methylation. 
perc_meth_bins_allmeth_10kb_wide <- perc_meth_bins_allmeth_10kb %>% 
  ungroup() %>%
  #tidyr::unite("gene_region", c("contig_gene", "start_gene_2kb", "end_gene_2kb"), sep = "_", remove = FALSE) %>%
  dplyr::select(population, sample, bin, mean_percMeth) %>%
  spread(bin, mean_percMeth) %>%
  tibble::column_to_rownames(var = "sample")
```

### Run Pst for 10kb bins (all with methylation data, not just bins with genetic+meth data)

```{r}
#Now run the following line and it will provide Pst estimates for every gene.
perc_meth_bins_allmeth_10kb_Pst <- Pst(perc_meth_bins_allmeth_10kb_wide) 

# Save 10kb Pst object to file for 
save(perc_meth_bins_allmeth_10kb_Pst, file="../analyses/methylation-filtered/R-objects/perc_meth_bins_allmeth_10kb_Pst")

# Check out Pst distribution 
hist(perc_meth_bins_allmeth_10kb_Pst$Pst_Values)
summary(perc_meth_bins_allmeth_10kb_Pst$Pst_Values)
mean(perc_meth_bins_allmeth_10kb_Pst$Pst_Values)
sd(perc_meth_bins_allmeth_10kb_Pst$Pst_Values)
nrow(perc_meth_bins_allmeth_10kb_Pst)

# format of dataframe that I will save 
perc_meth_bins_allmeth_10kb_Pst %>% separate(Quant_Varia, into=c("contig","start.bin", "end.bin"), sep = "_", remove = FALSE) %>% noquote()

# Write out Pst results 
# NOTE: this file ("Pst_bins_allmeth_10kb.tab") differs from the "Pst_bins_10kb.tab" file in that it contains Pst values for ALL 10kb bins, not just bins that also contain genetic data (many more bins!)
write.table(perc_meth_bins_allmeth_10kb_Pst %>% 
              separate(Quant_Varia, into=c("contig","start.bin", "end.bin"), sep = "_", remove = FALSE), 
            file = here::here("analyses", "methylation-filtered", "Pst_bins_allmeth_10kb.tab"), 
            sep = '\t', na = "NA", row.names = FALSE, col.names = TRUE, quote=FALSE)
```