---
title: "Figures"
author: "Laura H Spencer"
date: "7/13/2020"
output: html_document
---
```{r}
library(ggplot2)
library(spaa) #list2dist
library(ggpubr) #add correlation to plot
```

```{r}
library(tidyverse)
```




# Figure 3a
## PCA Figure, genetic data for only MBD samples 

```{r, fig.show='hide'}
#Create dataframe with sample treatment information, color & symbol scheme 
plotCustomization <- data.frame(sample = 1:18, population=c(
                                rep("HC", times=9), 
                                rep("SS", times=9)),
                                color=c(rep("firebrick3", times=9),
                                rep("dodgerblue3", times=9)),
                                symbol=c(rep(16, times=9),
                                rep(17, times=9))) 

# Check out genetic PCA axes 
load("../analyses/2bRAD/MethGen/MBD_pca_covMat.Robj") 

# calculate % variance explained by each PC axis (order from 1+): 
eig$val

# Generate figure 
par(mar = c(5, 5, 1, 1)) #Specify inner and outer margins
plot(1, type="n", xlab="", ylab="", xlim=c(-0.5, 0.5), ylim=c(-0.6, 0.8), axes=FALSE, ann=FALSE)
points(x=PC[,1], y=PC[,2], col = as.character(plotCustomization$color), pch = plotCustomization$symbol, cex = 2) #Add each sample. 
axis(side =  1, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add x-axis
mtext(side = 1, text = "PC 1 (9.5%)", line = 3, cex = 1.5) #Add x-axis label
axis(side =  2, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add y-axis
mtext(side = 2, text = "PC 2 (8.0%)", line = 3, cex = 1.5) #Add y-axis label
legend("topleft", 
       pch = c(16, 17), 
       legend = c("Hood Canal", "South Sound"), 
       col = c(c("firebrick3", "dodgerblue3")), 
       cex = 1.2, bty = "n") #Add legend 
```

## PCA Figure, genetic data for all samples 

```{r, fig.show='hide'}
# Load DF with PCA axis info 
PCA.allgen <- read.table(file = "../analyses/2bRAD/PopGen/All_pca_covMat.tsv")

# Find min and max for PC axes 1 and 2
summary(PCA.allgen$PC1) #min= --0.1384, max=0.137 
summary(PCA.allgen$PC2) #min= -0.1643, max=0.2098

# Generate a DF to define color and shape of each point 
plotCustomization <- rownames(PCA.allgen) %>% str_extract(c("HC|SS")) %>% as.tibble() %>% 
  rename(population = value) %>%
  mutate(color=ifelse(population=="HC", "firebrick3", "dodgerblue3")) %>%
  mutate(symbol=ifelse(population=="HC", 16, 17)) 

# Generate figure 
pdf(NULL)
dev.control(displaylist="enable")



#png("../analyses/2bRAD/PopGen/PCA_covmat_all.png")
par(mar = c(5, 5, 1, 1)) #Specify inner and outer margins
plot(1, type="n", xlab="", ylab="", xlim=c(-0.16, 0.16), ylim=c(-0.2, 0.22), axes=FALSE, ann=FALSE)
points(x=PCA.allgen$PC1, y=PCA.allgen$PC2, col = as.character(plotCustomization$color), 
       pch = plotCustomization$symbol, cex = 1.2) #Add each sample. 
axis(side =  1, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add x-axis
mtext(side = 1, text = "PC 1 (4.5%)", line = 3, cex = 1.5) #Add x-axis label
axis(side =  2, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add y-axis
mtext(side = 2, text = "PC 2 (3.38%)", line = 3, cex = 1.5) #Add y-axis label
legend(x = -0.13, y=0.17, 
       pch = c(16, 17), 
       legend = c("Hood Canal", "South Sound"), 
       col = c(c("firebrick3", "dodgerblue3")), 
       cex = 1.3, bty = "n") #Add legend 

p1.base <- recordPlot()
invisible(dev.off())
#dev.off()
```

## PCA Figure, all methylated loci after filtering 

```{r, fig.show='hide'}
load("../analyses/methylation-filtered/R-objects/meth_filter") #load filtered meth data object 

plotCustomization <- data.frame(sample = 1:18, population=c(
                                rep("HC", times=9), 
                                rep("SS", times=9)),
                                color=c(rep("firebrick3", times=9),
                                rep("dodgerblue3", times=9)),
                                symbol=c(rep(16, times=9),
                                rep(17, times=9))) 

PCA.filtered <- PCASamples(meth_filter, obj.return = TRUE) #Run a PCA analysis on percent methylation for all samples. methylKit uses prcomp to create the PCA matrix
summary(PCA.filtered) #Look at summary statistics. The first PC explains 9.879% of variation, the second PC explains 8.51% of variation


# Create figure 
par(mar = c(5, 5, 1, 1)) #Specify inner and outer margins
PCA.figure <- ordiplot(PCA.filtered, choices = c(1, 2), type = "none", display = "sites", cex = 0.5, xlab = "", ylab = "", xaxt = "n", yaxt = "n") #Use ordiplot to create base biplot. Do not add any points
points(PCA.figure, "sites", col = as.character(plotCustomization$color), pch = plotCustomization$symbol, cex = 2) #Add each sample. Darker samples are ambient, lighter samples are elevated pCO2
#Add multiple white boxes on top of the default black box to manually change the color
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
#ordiellipse(PCA.filtered, plotCustomization$population, show.groups = "HC", col = "firebrick3") #Add confidence ellipse around the samples in elevated pCO2
#ordiellipse(PCA.filtered, plotCustomization$population, show.groups = "SS", col = "dodgerblue3") #Add confidence ellipse around the samples in ambient pCO2
axis(side =  1, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add x-axis
mtext(side = 1, text = "PC 1 (9.87%)", line = 3, cex = 1.5) #Add x-axis label
axis(side =  2, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add y-axis
mtext(side = 2, text = "PC 2 (8.51%)", line = 3, cex = 1.5) #Add y-axis label
legend("topleft", 
       pch = c(16, 17), 
       legend = c("Hood Canal", "South Sound"), 
       col = c(c("firebrick3", "dodgerblue3")), 
       cex = 1.2, bty = "n") #Add a legend with information about ambient and elevated samples
```

## PCA Figure, DMLs only 

```{r, fig.show='hide'}
load("../analyses/DMLs/R-objects/dml25_counts")

PCA.DMLs <- PCASamples(dml25_counts, obj.return = TRUE, sd.filter = T) #Run a PCA analysis on percent methylation for all samples. methylKit uses prcomp to create the PCA matrix
summary(PCA.DMLs) #Look at summary statistics. The first PC explains 36.3% of variation, the second PC explains 8.47% of variation

par(mar = c(5, 5, 1, 1)) #Specify inner and outer margins
PCA.figure <- ordiplot(PCA.DMLs, choices = c(1, 2), type = "none", display = "sites", cex = 0.5, xlab = "", ylab = "", xaxt = "n", yaxt = "n") #Use ordiplot to create base biplot. Do not add any points
points(PCA.figure, "sites", col = as.character(plotCustomization$color), pch = plotCustomization$symbol, cex = 2) #Add each sample. 
#Add multiple white boxes on top of the default black box to manually change the color
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
#ordiellipse(PCA.DMLs, plotCustomization$population, show.groups = "HC", col = "firebrick3") #Add confidence ellipse around the samples in elevated pCO2
#ordiellipse(PCA.DMLs, plotCustomization$population, show.groups = "SS", col = "dodgerblue3") #Add confidence ellipse around the samples in ambient pCO2
axis(side =  1, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add x-axis
mtext(side = 1, text = "PC 1 (36.3%)", line = 3, cex = 1.5) #Add x-axis label
axis(side =  2, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add y-axis
mtext(side = 2, text = "PC 2 (8.47%)", line = 3, cex = 1.5) #Add y-axis label
legend("topright", 
       pch = c(16, 17), 
       legend = c("Hood Canal", "South Sound"), 
       col = c(c("firebrick3", "dodgerblue3")), 
       cex = 1.3, bty = "n") #Add a legend 
```

## Heatmap of DMLs 

```{r}
load("../analyses/DMLs/R-objects/DMLs_heatmap")

#ggplotly(
ggplot(DMLs_heatmap, aes(sample, chr_start, fill= percMeth)) + xlab("Sample") + geom_tile(na.rm = T) +
  scale_y_discrete(limits=(DML.ratios[order(DML.ratios$ratio_HC.SS),]$chr_start)) + 
  #scale_fill_viridis(discrete=FALSE) 
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) + 
  scale_fill_distiller(palette = "YlGnBu", direction = 1) #)
  #scale_fill_gradient(low="gray5", high="white")
```


## Check out overall distribution of filtered methylation data by population 
Any trends? Looks like HC has more instances of ~100% methylation with a lower second peak at around ~92%. Compared to SS that has a lower peak at ~100% and a second around 95%. 

```{r}
load(file="../analyses/methylation-filtered/R-objects/meth_filter_reshaped")
meth_filter_reshaped %>% 
  ggplot(aes(x=population, y=percMeth, fill=population)) +
  geom_violin(width=1.1, alpha=0.8) + theme_minimal() +
  ylab("% CpG Methylation") + theme(axis.title.x = element_blank(), legend.position = "none",
                                    text = element_text(size=18)) +
  scale_x_discrete(labels=c("HC" = "Hood Canal\nPopulation", "SS" = "South Sound\nPopulation")) + 
  scale_fill_manual(values=c("firebrick3", "dodgerblue3"))
```


# Figure 4a
## Correlation of Epigenetic/Genetic Distance  
```{r}
#load("../analyses/2bRAD/MethGen/dist_allmeth_snp.Robj")
ep10 <-read.csv("../analyses/methylation-filtered/dist.manhat.csv",header = T)
ep10 <- ep10[,c("SampNum.row","SampNum.col","dist.manh")]
ep10 <- as.matrix(list2dist(ep10))
mbdorder = c("hc1_2","hc1_4","hc2_15","hc2_17","hc3_1","hc3_5","hc3_7","hc3_10","hc3_11",
        "ss2_9","ss2_14","ss2_18","ss3_3","ss3_14","ss3_15","ss3_16","ss3_20","ss5_18")
ep10 <- ep10[mbdorder, mbdorder]

gen <- read.table("../analyses/2bRAD/PopGen/HCSS_Afilt32m70_01_mbd.dist", row.names = 1, header=T)

df <- data.frame( gen=gen[lower.tri(gen)], ep=ep10[ lower.tri(ep10)])

sp <- ggplot(df, aes(x=gen,y=ep)) + geom_point(shape=19) +    # Use hollow circles
    geom_smooth(method=lm) +
  theme_linedraw()+
theme(
axis.title.x = element_text( size=15),
axis.title.y = element_text(size=15),
panel.grid.major = element_blank(),
panel.grid.minor = element_blank()
) +
  xlab("Genetic distance") + ylab("Methylation distance")  +
  scale_y_continuous(breaks = c(2250000,3250000),labels = c("225","325"))+
   stat_cor(method = "pearson", label.x = 0.115, label.y = 3250000,size=5)
  
sp
```


# Figure 4b
## Correlation of Epigenetic/Genetic Distance (DML) 
```{r}
#load("../analyses/2bRAD/MethGen/dist_allmeth_snp.Robj")
ep10 <-read.csv("../analyses/methylation-filtered/dist.manhat.DMLs.csv",header = T)
ep10 <- ep10[,c("SampNum.row","SampNum.col","dist.manh")]
ep10 <- as.matrix(list2dist(ep10))
mbdorder = c("hc1_2","hc1_4","hc2_15","hc2_17","hc3_1","hc3_5","hc3_7","hc3_10","hc3_11",
        "ss2_9","ss2_14","ss2_18","ss3_3","ss3_14","ss3_15","ss3_16","ss3_20","ss5_18")
ep10 <- ep10[mbdorder, mbdorder]

df <- data.frame( gen=gen[lower.tri(gen)], ep=ep10[ lower.tri(ep10)])

sp <- ggplot(df, aes(x=gen,y=ep)) + geom_point(shape=19) +    # Use hollow circles
    geom_smooth(method=lm) +
  theme_linedraw()+
theme(
axis.title.x = element_text( size=15),
axis.title.y = element_text(size=15),
panel.grid.major = element_blank(),
panel.grid.minor = element_blank()
) +
  xlab("Genetic distance") + ylab("Methylation distance (DMLs)")  +
  scale_y_continuous(breaks = c(50000,200000),labels = c("5","20"))+
   stat_cor(method = "pearson", label.x = 0.115, label.y = 190000,size=5,p.digits = 0.001)
  
sp
```

# Figure 5
## Correlation of Fst and Pst (10kb bins)  
```{r}
fst <- read.csv("../analyses/2bRAD/MethGen/HCSS_sfsm70_Per10kbFst.csv")
id <- paste(fst$contig,fst$start,fst$end, sep = "_")
fst <- cbind(id,fst)

pst10kb <- read.csv("../analyses/methylation-filtered/Pst_bins_10kb.tab", sep="\t",header=T,stringsAsFactors = F)
colnames(pst10kb) <- c("id","contig","start","end","pst","pst.lowCI","pst.highCI")

both = merge(pst10kb,fst, by ="id")

sp <- ggplot(both, aes(x=fst01,y=pst)) + geom_point(shape=19) +    # Use hollow circles
    geom_smooth(method=lm) +
  theme_linedraw()+
theme(
axis.title.x = element_text( size=15),
axis.title.y = element_text(size=15),
panel.grid.major = element_blank(),
panel.grid.minor = element_blank()
) +
  xlab("Fst") + ylab("Pst")  +
  #scale_y_continuous(breaks = c(9000,15000),labels = c("9","15"))+
   stat_cor(method = "pearson", label.x = 0.35, label.y = 0.87,size=5,p.digits = 0.001)
  
sp
```

## Correlation of methylation PC2 and SNP PC1
```{r}
ep <- read.table("../analyses/methylation-filtered/PC-scores-filtered-methylation.tab",
                 header=T, sep="\t",row.names = "sample")
#add sample names
key = read.csv("../data/sample-key.csv",colClasses = c("character","character"))
samples = rownames(ep)
mapdf <- data.frame(old=key$MBD.FILENAME,new=key$SAMPLE)
rownames(ep) <- mapdf$new[match(samples,mapdf$old)]

gen <- read.table("../analyses/2bRAD/MethGen/MBD_pca_covMat.tsv")
mbdorder = c("hc1_2","hc1_4","hc2_15","hc2_17","hc3_1","hc3_10","hc3_11","hc3_5","hc3_7",
  "ss2_14","ss2_18","ss2_9","ss3_14","ss3_15","ss3_16","ss3_20","ss3_3","ss5_18")
rownames(gen) <- mbdorder
#make sure ep and gen in same order
ep <- ep[mbdorder,]

both <- merge(ep[,"PC2"],gen[,"PC1"],by="row.names",sort=F)
Pop <- as.factor(c(rep("Hood Canal",9),rep("South Sound",9)))
both <- cbind(both,Pop)
sp <- ggplot(both, aes(x=y,y=x)) + 
  geom_point(aes(shape=Pop,color=Pop),size=3) +
  scale_shape_manual(values=c(16,17)) + 
  scale_color_manual(values=c("firebrick3", "blue3")) +
    geom_smooth(method=lm) +
  theme_linedraw()+
theme(
axis.title.x = element_text( size=15),
axis.title.y = element_text(size=15),
panel.grid.major = element_blank(),
panel.grid.minor = element_blank()
) +
  xlab("PC1 SNPs") + ylab("PC2 Methylation")  +
   stat_cor(method = "pearson", label.x = 0, label.y = 200,size=4,p.digits = 0.001)
  
sp
#ggsave(sp, file = "../analyses/2bRAD/MethGen/methPC2_SNPPC1.png", dpi = 800)
```
## Plotting mQTLs
```{r}
load("../analyses/2bRAD/mQTL_5x/Contig54624.19738.19738.df")
loci = "Contig54624.19738.19738"
p <- ggplot(df, aes(value, percMeth, fill=pop))  +
        geom_boxplot(alpha=0.7) +
  geom_dotplot(binaxis='y', stackdir='center',dotsize=0.8, position = position_dodge(0.75)) +
        theme_linedraw() + ggtitle("A.") +
  xlab("SNP Genotype") + ylab("% Methylation") + 
  theme(legend.title = element_blank(),
        legend.position = c(0.82,0.85),
  legend.text = element_text(size=12),                                                 
axis.title.x = element_text( size=15),
axis.title.y = element_text(size=15),
panel.grid.major = element_blank(),
panel.grid.minor = element_blank(),
plot.title = element_text(size=19)) +
  scale_fill_discrete(labels=c("Hood Canal","South Sound"))
p
```
```{r}
load("../analyses/2bRAD/mQTL_5x/Contig60108.5780.5780.df")
p <- ggplot(df, aes(value, percMeth, fill=pop))  +
        geom_boxplot(alpha=0.7) +
  geom_dotplot(binaxis='y', stackdir='center',dotsize=0.8, position = position_dodge(0.75)) +
        theme_linedraw() + ggtitle("B.") +
  xlab("SNP Genotype") + ylab("% Methylation") + 
  theme(legend.title = element_blank(),
        legend.position = c(0.82,0.85),
  legend.text = element_text(size=12),                                                 
axis.title.x = element_text( size=15),
axis.title.y = element_text(size=15),
panel.grid.major = element_blank(),
panel.grid.minor = element_blank(),
plot.title = element_text(size=19)) +
  scale_fill_discrete(labels=c("Hood Canal","South Sound"))
p
```
