---
title: "Supplementary"
author: "Katherine Silliman, Laura H Spencer, Steven Roberts"
date: "1/26/2022"
output: html_document
---
## Supplementary 
### _Epigenetic and genetic population structure is coupled in a marine invertebrate_
### Katherine Silliman, Laura H Spencer, Steven B Roberts

```{r, include=FALSE}
# Load libraries
list.of.packages <- c("spaa", "ggpubr", "tidyverse", "methylKit", "gridGraphics", "gridExtra", "vegan", "scales", "pryr", "corrplot") #add libraries here 

# Install packages if needed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load packages 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})
```


### Supplemental Figure 1
#### Frequency distribution of % methylation in _O. lurida_ prior to filtering (left) and after filtering for loci with at minimum 5x coverage (right).

```{r, warning=FALSE}
load(file="../analyses/methylation-characteristics/R-objects/allmeth")
load(file="../analyses/methylation-characteristics/R-objects/allmeth.5x")

# Frequency distribution of % methylation 
supp1a %<a-% hist(allmeth$meth_perc, col="gray85", 
                  xlab="% methylation per base", cex.main=.9,
     main="% methylation, no coverage minimum\n(all samples concatenated)") 

# Frequency distribution of % methylation 
supp1b %<a-% hist(allmeth.5x$meth_perc, col="gray85", 
     xlab="% methylation per base", cex.main=.9,
     main="% methylation, 5x minimum coverage\n(all samples concatenated)") 

supp.figure01 %<a-% {
  split.screen(c(1, 2))
  screen(1)
  supp1a
  screen(2)
  supp1b
  close.screen(all=TRUE)
}

supp.figure01

# tiff(file = "../figures/supp.figure01.tiff", width=3000, height=1250, res=250) 
# supp.figure01
# dev.off()
```

### Supplemental Figure 2
#### Top: Table showing the % of all CpG loci and methylated loci that overlap with each genomic feature. Bottom: A correlation plot, showing residuals from chi-squared tests of homogeneity on the distribution of methylated loci and CpG loci that intersect with features. Blue=positive associations, red=negative associations, and size of circles represents the absolute value of each correlation coefficient. 

| Feature | % of all CpGs in genome | % of all methylated loci | Ratio of  % meth:%CpG |
|---|---|---|---|
| Exon | 3.99% | 14.71% | 3.69 |
| Intron | 13.94% | 19.76% | 1.42 |
| 5’ flanking region (-2kb) | 3.91% | 4.64% | 1.19 |
| 3’ flanking region (+2kb) | 3.92% | 4.66% | 1.19 |
| Transposable Elements | 14.93% | 13.83% | 0.93 |
| Unknown genome regions | 56.01% | 32.25% | 0.58 |

```{r, warning=FALSE}
load(file="../analyses/methylation-characteristics/R-objects/loci.summary")

meth.chisq <- methdata.summary %>% filter(feature=="exon" | feature=="intron" | feature=="2kbflank-up" | feature=="2kbflank-down" | feature=="TE" | feature=="unknown") %>% 
  column_to_rownames("feature") %>% dplyr::select(CpGs, methylated) %>% chisq.test() 

supp2 %<a-% corrplot(meth.chisq$residuals, is.cor = FALSE, method="circle", cl.pos = 'n')

supp2

# tiff(file = "../figures/supp.figure02.tiff", width=750, height=1250, res=250) 
# supp2
# dev.off()
```

### Supplemental Figure 3
#### Relative location of methylation within gene bodies in _Ostrea lurida_. For this analysis methylated loci were designated as those with >50% methylation when averaged across all MBD18 samples.

```{r, warning=FALSE}
load(file="../analyses/methylation-filtered/meth_filter_genes") 

# Histogram showing frequency of methylated loci (min 50% meth) data across gene (relative location)
(supp.3a <- meth_filter_genes %>% 
  group_by(gene, gen_location) %>% 
  summarise(mean_meth=mean(percMeth)) %>%
  filter(mean_meth>50) %>%
  ggplot(., aes(x=gen_location)) +
  stat_bin(bins=150, col="gray30", fill="lavenderblush3") +
  xlab("Relative location along gene") + 
  ylab("Frequency of methylation data") + 
  ggtitle("Methylation frequency relative to gene location") +
  theme_minimal())

(supp.3b <- meth_filter_genes %>% 
  ungroup() %>%
  group_by(gen_location_round) %>% 
  summarise(mean_meth=mean(percMeth, na.rm=TRUE), sd_meth=sd(percMeth, na.rm=TRUE)) %>%
  ggplot(., aes(x=gen_location_round, y=mean_meth)) +
  geom_bar(stat="identity", col="gray30", fill="lavenderblush3") +
  # geom_errorbar(aes(ymin=mean_meth-sd_meth, ymax=mean_meth+sd_meth), width=.02,
  #                position=position_dodge(.9), ) +
  scale_y_continuous(limits = c(0, 100)) +
  xlab("Relative location along gene") +
  ylab("Mean % methylation") +
  ggtitle("% methylation by relative gene location") +
  theme_minimal())

# Plot methylation variance by gene location  
(supp.3c <- meth_filter_genes %>% 
  ungroup() %>%
  group_by(gen_location_round) %>% 
  summarise(var_meth=var(percMeth, na.rm=TRUE)) %>%
  ggplot(., aes(x=gen_location_round, y=var_meth)) +
  geom_bar(stat="identity", col="gray30", fill="lavenderblush3") +
  xlab("Relative location along gene") +
  ylab("Variance in % methylation") +
  ggtitle("Variance in % methylation by relative gene location") +
  theme_minimal())

supp.figure03 <- ggarrange(supp.3a, supp.3b, supp.3c, ncol=1, nrow=3)

```
### Supplemental Table 1
#### Biological functions that are enriched in the _Ostrea lurida_ genome.

| GO Term | Biological Process | PValue | FDR | GO slim |
|---|---|---|---|---|
| GO:0007067 | mitotic nuclear division | 0 | 0.37 | cell cycle and proliferation |
| GO:0007049 | cell cycle | 0 | 0.66 | cell cycle and proliferation |
| GO:0051321 | meiotic cell cycle | 0.05 | 1 | cell cycle and proliferation |
| GO:0000082 | G1/S transition of mitotic cell cycle | 0.07 | 1 | cell cycle and proliferation |
| GO:0000086 | G2/M transition of mitotic cell cycle | 0.09 | 1 | cell cycle and proliferation |
| GO:0007067 | mitotic nuclear division | 0 | 0.37 | cell organization and biogenesis |
| GO:0016569 | covalent chromatin modification | 0.01 | 1 | cell organization and biogenesis |
| GO:0060271 | cilium morphogenesis | 0.01 | 1 | cell organization and biogenesis |
| GO:0042384 | cilium assembly | 0.01 | 1 | cell organization and biogenesis |
| GO:0007030 | Golgi organization | 0.02 | 1 | cell organization and biogenesis |
| GO:0030030 | cell projection organization | 0.05 | 1 | cell organization and biogenesis |
| GO:0007005 | mitochondrion organization | 0.05 | 1 | cell organization and biogenesis |
| GO:0001822 | kidney development | 0.02 | 1 | developmental processes |
| GO:0001843 | neural tube closure | 0.07 | 1 | developmental processes |
| GO:0006281 | DNA repair | 0 | 0.08 | DNA metabolism |
| GO:0006260 | DNA replication | 0 | 0.37 | DNA metabolism |
| GO:0006310 | DNA recombination | 0.04 | 1 | DNA metabolism |
| GO:0006302 | double-strand break repair | 0.08 | 1 | DNA metabolism |
| GO:0006284 | base-excision repair | 0.09 | 1 | DNA metabolism |
| GO:0051301 | cell division | 0 | 0.12 | other biological processes |
| GO:0007018 | microtubule-based movement | 0.02 | 1 | other biological processes |
| GO:0042254 | ribosome biogenesis | 0.05 | 1 | other biological processes |
| GO:0016032 | viral process | 0.05 | 1 | other biological processes |
| GO:0043547 | positive regulation of GTPase activity | 0.05 | 1 | other biological processes |
| GO:0000910 | cytokinesis | 0.06 | 1 | other biological processes |
| GO:0032092 | positive regulation of protein binding | 0.08 | 1 | other biological processes |
| GO:0008104 | protein localization | 0.09 | 1 | other biological processes |
| GO:0031047 | gene silencing by RNA | 0.01 | 1 | other metabolic processes |
| GO:0006511 | ubiquitin-dependent protein catabolic process | 0 | 0.12 | protein metabolism |
| GO:0006412 | translation | 0 | 0.37 | protein metabolism |
| GO:0050821 | protein stabilization | 0.01 | 1 | protein metabolism |
| GO:0006468 | protein phosphorylation | 0.01 | 1 | protein metabolism |
| GO:0006413 | translational initiation | 0.03 | 1 | protein metabolism |
| GO:0000209 | protein polyubiquitination | 0.06 | 1 | protein metabolism |
| GO:0042787 | protein ubiquitination involved in ubiquitin-dependent protein catabolic process | 0.08 | 1 | protein metabolism |
| GO:0016567 | protein ubiquitination | 0.09 | 1 | protein metabolism |
| GO:0000398 | mRNA splicing, via spliceosome | 0 | 0.53 | RNA metabolism |
| GO:0006397 | mRNA processing | 0 | 1 | RNA metabolism |
| GO:0006364 | rRNA processing | 0.01 | 1 | RNA metabolism |
| GO:0006396 | RNA processing | 0.01 | 1 | RNA metabolism |
| GO:0008380 | RNA splicing | 0.02 | 1 | RNA metabolism |
| GO:0006366 | transcription from RNA polymerase II promoter | 0.05 | 1 | RNA metabolism |
| GO:0008033 | tRNA processing | 0.06 | 1 | RNA metabolism |
| GO:0006355 | regulation of transcription, DNA-templated | 0.09 | 1 | RNA metabolism |
| GO:0006974 | cellular response to DNA damage stimulus | 0 | 0.05 | stress response |
| GO:0006281 | DNA repair | 0 | 0.08 | stress response |
| GO:0006302 | double-strand break repair | 0.08 | 1 | stress response |
| GO:0006284 | base-excision repair | 0.09 | 1 | stress response |
| GO:0015031 | protein transport | 0 | 0.23 | transport |
| GO:0006886 | intracellular protein transport | 0 | 0.81 | transport |
| GO:0006888 | ER to Golgi vesicle-mediated transport | 0 | 0.96 | transport |
| GO:0016192 | vesicle-mediated transport | 0.02 | 1 | transport |
| GO:0006406 | mRNA export from nucleus | 0.04 | 1 | transport |
| GO:0006606 | protein import into nucleus | 0.09 | 1 | transport |
| GO:0042147 | retrograde transport, endosome to Golgi | 0.1 | 1 | transport |
| GO:0098609 | cell-cell adhesion | 0.04 | 1 | NA |
| GO:0070936 | protein K48-linked ubiquitination | 0.07 | 1 | NA |
| GO:0003341 | cilium movement | 0.07 | 1 | NA |

### Supplemental Figure 4
#### Heat map of loci that are differentially methylated (DMLs) between the two populations.

```{r, warning=FALSE}
load("../analyses/DMLs/R-objects/DMLs_heatmap")
load("../analyses/DMLs/R-objects/DML.ratios")

DMLs_heatmap %>% dplyr::select(chr, start, end) %>% unique()
#ggplotly(
ggplot(DMLs_heatmap, aes(sample, chr_start, fill= percMeth)) + xlab("Sample") + geom_tile(na.rm = T) +
  scale_y_discrete(limits=(DML.ratios[order(DML.ratios$ratio_HC.SS),]$chr_start)) + 
  #scale_fill_viridis(discrete=FALSE) 
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) + 
  scale_fill_distiller(palette = "YlGnBu", direction = 1) +
coord_cartesian(ylim = c(0, NA), clip = "off") + xlab("") +
  annotate("segment", x = 1, xend = 9, y = -225, yend = -225, colour="firebrick3") +
    annotate("segment", x = 10, xend = 18, y = -225, yend = -225, colour="dodgerblue3") +
  annotate("text", x = 5, y = -350, colour="firebrick3", label="Hood Canal") +
    annotate("text", x = 14, y = -350, colour="dodgerblue3", label="South Sound")#)#)
  #scale_fill_gradient(low="gray5", high="white")
```

### Supplemental Figure 5
#### PCA of methylation data using DMLs only 

```{r, warning=FALSE}
load(file="../analyses/DMLs/R-objects/PCA.figure")

PC.coord <- PCA.figure$sites %>% as_tibble() %>% 
  cbind(Population=c(rep("Hood Canal", times=9), rep("South Sound", times=9))) %>% 
  mutate(Population=as.factor(Population))

(supp5 <- ggplot(PC.coord, aes(x=PC1,y=PC2)) + 
  geom_point(aes(shape=Population,color=Population),size=4) +
  scale_shape_manual(values=c(16,17)) + 
  scale_color_manual(values=c("firebrick3", "blue3")) +
  theme_linedraw()+
theme(
axis.title.x = element_text( size=15),
axis.title.y = element_text(size=15),
panel.grid.major = element_blank(),
panel.grid.minor = element_blank(),
legend.position = "right"
) + 
    xlab("PC 1 (36.3%)") + ylab("PC 2 (8.47%%)") +
      xlim(-29,30) + ylim(-21,28))
```

### Supplemental Figure 6
#### The % difference between each population’s mean methylation for differentially methylated loci, which highlights that both populations have a similar number of hyper/hypo-methylated loci. Loci are sorted by % methylation difference.

```{r, warning=FALSE}
load("../analyses/DMLs/R-objects/DML.ratios")

DML.ratios %>% 
           mutate(hypermethylated=ifelse(diff_HC.SS>0, "HC", "SS")) %>%
  ggplot(aes(x = chr_start, y = diff_HC.SS, fill=hypermethylated)) +
  geom_bar(stat = "identity", width=0.41) +
    scale_x_discrete(limits=(DML.ratios[order(DML.ratios$diff_HC.SS),]$chr_start)) + 
  scale_y_continuous(limits=c(-80,80), breaks=c(-75, -50, -25, 0),
                       labels=c("-75"="75%", "-50"="50%", "-25"="25%", "0"="0%"),
                     sec.axis = sec_axis(trans=~.*1, breaks=c(75, 25, 50, 0),
                       labels=c("-75"="75%", "-50"="50%", "-25"="25%", "0"="0%")))  +
    theme(axis.text.y=element_blank(), axis.ticks.y = element_blank(), 
          legend.position = "none", text = element_text(size=14, color="gray30"),
          panel.background = element_blank(),
          panel.border = element_rect(colour = "gray30", fill=NA, size=.4)) +
    #ggtitle("DMLs, showing % methylation difference among populations") +
    scale_fill_manual(values=c("firebrick3", "dodgerblue3")) +
    # annotate(geom="text", x=-20, y=-65, label="Loci hypermethylated in\nSouth Sound", color="dodgerblue3") +
    # annotate(geom="text", x=260, y=62, label="Loci hypermethylated in\nHood Canal", color="firebrick3") +
  ylab("% methylation difference among populations") + xlab("Locus") + coord_flip()
```
### Supplemental Table 2
#### GO terms and GO Slim terms of biological functions enriched in DMGs & DMLs. EASE scores shown. FDR for all terms is 1.0. 

| Go Slim | Term | Function | DMG | DML |
|---|---|---|---|---|
| cell adhesion | GO:0007155 | cell adhesion | 0.012 |   |
| cell adhesion | GO:0007156 | homophilic cell adhesion via plasma membrane adhesion molecules | 0.026 | 0.00091 |
| cell adhesion | GO:0016339 | calcium-dependent cell-cell adhesion via plasma membrane cell adhesion molecules |   | 0.083 |
| cell organization and biogenesis | GO:0007015 | actin filament organization | 0.014 |   |
| cell organization and biogenesis | GO:0048675 | axon extension | 0.047 |   |
| cell organization and biogenesis | GO:0007411 | axon guidance | 0.063 |   |
| cell organization and biogenesis | GO:0000904 | cell morphogenesis involved in differentiation | 0.095 |   |
| cell organization and biogenesis | GO:0030198 | extracellular matrix organization | 0.033 |   |
| cell organization and biogenesis | GO:0032836 | glomerular basement membrane development | 0.046 |   |
| cell organization and biogenesis | GO:0007040 | lysosome organization | 0.047 |   |
| cell organization and biogenesis | GO:0031023 | microtubule organizing center organization | 0.095 |   |
| cell organization and biogenesis | GO:0051259 | protein oligomerization | 0.098 |   |
| cell organization and biogenesis | GO:0048841 | regulation of axon extension involved in axon guidance | 0.033 |   |
| cell organization and biogenesis | GO:0045214 | sarcomere organization | 0.069 | 0.019 |
| cell organization and biogenesis | GO:0055003 | cardiac myofibril assembly |   | 0.082 |
| cell organization and biogenesis | GO:0032438 | melanosome organization |   | 0.021 |
| cell organization and biogenesis | GO:0045332 | phospholipid translocation |   | 0.0058 |
| death | GO:0008625 | extrinsic apoptotic signaling pathway via death domain receptors | 0.095 |   |
| developmental processes | GO:0048675 | axon extension | 0.047 |   |
| developmental processes | GO:0007411 | axon guidance | 0.063 |   |
| developmental processes | GO:0032836 | glomerular basement membrane development | 0.046 |   |
| developmental processes | GO:0048841 | regulation of axon extension involved in axon guidance | 0.033 |   |
| developmental processes | GO:0045214 | sarcomere organization | 0.069 | 0.019 |
| developmental processes | GO:0007423 | sensory organ development | 0.095 |   |
| developmental processes | GO:0055003 | cardiac myofibril assembly |   | 0.082 |
| developmental processes | GO:0060429 | epithelium development |   | 0.071 |
| developmental processes | GO:0040027 | negative regulation of vulval development |   | 0.082 |
| developmental processes | GO:0021942 | radial glia guided migration of Purkinje cell |   | 0.082 |
| developmental processes | GO:0050767 | regulation of neurogenesis |   | 0.0058 |
| developmental processes | GO:0060438 | trachea development |   | 0.082 |
| developmental processes | GO:0001570 | vasculogenesis |   | 0.047 |
| other biological processes | GO:0016477 | cell migration | 0.029 | 0.032 |
| other biological processes | GO:0007281 | germ cell development | 0.048 |   |
| other biological processes | GO:0019915 | lipid storage | 0.043 |   |
| other biological processes | GO:0048477 | oogenesis | 0.046 |   |
| other biological processes | GO:0040008 | regulation of growth | 0.032 |   |
| other biological processes | GO:0042254 | ribosome biogenesis | 0.092 |   |
| other biological processes | GO:0019233 | sensory perception of pain | 0.016 |   |
| other biological processes | GO:0006879 | cellular iron ion homeostasis |   | 0.047 |
| other biological processes | GO:0050982 | detection of mechanical stimulus |   | 0.082 |
| other biological processes | GO:0050801 | ion homeostasis |   | 0.082 |
| other biological processes | GO:0007017 | microtubule-based process |   | 0.082 |
| other biological processes | GO:0032465 | regulation of cytokinesis |   | 0.057 |
| other biological processes | GO:0006941 | striated muscle contraction |   | 0.082 |
| other metabolic processes | GO:0042157 | lipoprotein metabolic process | 0.095 |   |
| other metabolic processes | GO:0010508 | positive regulation of autophagy | 0.063 |   |
| other metabolic processes | GO:0010923 | negative regulation of phosphatase activity |   | 0.028 |
| protein metabolism | GO:0031398 | positive regulation of protein ubiquitition | 0.098 | 0.093 |
| protein metabolism | GO:0016567 | protein ubiquitition |   | 0.0033 |
| protein metabolism | GO:0042787 | protein ubiquitition involved in ubiquitin-dependent protein catabolic process |   | 0.003 |
| signal transduction | GO:0046426 | negative regulation of JAK-STAT cascade |   | 0.071 |
| stress response | GO:0042594 | response to starvation |   | 0.03 |
| transport | GO:0006895 | Golgi to endosome transport | 0.043 | 0.019 |
| transport | GO:0034220 | ion transmembrane transport | 0.086 | 0.00049 |
| transport | GO:0008333 | endosome to lysosome transport |   | 0.083 |
| transport | GO:0045332 | phospholipid translocation |   | 0.0058 |
| NA  | GO:0044331 | cell-cell adhesion mediated by cadherin | 0.033 | 0.071 |
| NA  | GO:0072015 | glomerular visceral epithelial cell development | 0.095 |   |
| NA  | GO:0086010 | membrane depolarization during action potential | 0.095 |   |
| NA  | GO:1903955 | positive regulation of protein targeting to mitochondrion | 0.0082 |   |
| NA  | GO:0038061 | NIK/NF-kappaB sigling |   | 0.071 |
| NA  | GO:0090175 | regulation of establishment of plar polarity |   | 0.082 |