{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 05-MACAU \n",
    "\n",
    "In this notebook I use MACAU - Mixed model Association for Count data via data AUgmentation - to assess the influence of Olympia oyster size/weight on DNA methylation, while controlling for relatedness. I leverage sequence data from bisulfite treated DNA from 2 Olympia oyster populations - Dabob Bay (in Hood Canal), and Oyster Bay (in South Puget Sound). \n",
    "\n",
    "MACAU runs on Lenox. The software is available on [Zhou Lab website](http://www.xzlab.org/software.html). The R program seems to have some issues, so I'm using the binary version installed by Sam on Roadrunner. The [MACAU user manual](http://www.xzlab.org/software/macau/MACAUmanual.pdf) is the basis for this analysis. \n",
    "\n",
    "### Input files \n",
    "\n",
    "MACAU inputs for this type of analysis include: \n",
    "\n",
    "(1) Methylated read counts (raw counts, not percentages)\n",
    "(2) Total read counts (raw counts)\n",
    "(3) Relatedness matrix\n",
    "(4) Predictor variable, or covariates \n",
    "\n",
    "First, I download all the files I'll need for MACAU. The files were generated as follows: \n",
    "\n",
    "Both read count files - (1) and (2) above - were generated by Laura Spencer in RStudio in the [04-raw-count-files.Rmd notebook](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/code/04-raw-count-files.Rmd), using the MethylKit object that Steven Roberts generated in MethylKit in the [01-methylkit.Rmd notebook](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/code/01-methylkit.Rmd). Note that in both files, counts from + and - strand were combined. These matricese were created from a MethylKit object that had been filtered for all loci with minimum of 10x coverage and maximum of 100x coverage, and with loci that were present in 8 of the 9 samples per population.\n",
    "\n",
    "- [counts-filtered-total.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/counts-filtered-total.txt?raw=true)\n",
    "- [counts-filtered-meth.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/counts-filtered-meth.txt?raw=true)\n",
    "\n",
    "The relatedness matrix - (3) - was generated by Katherine Silliman from 2bRad data, using SNPs and the program ANGSD. Check out her notebook entry, [SOS_angsd.ipynb](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/2bRAD/SOS_angsd.ipynb). \n",
    "- [HCSSmbdsamples_rab.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/3f81bc8fdb8bc02f2a8aa585613fbc1b27b33333/analyses/2bRAD/HCSSmbdsamples_rab.txt) - generated using only HC/SS samples  \n",
    "- [mbdsamples_rab.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/2bRAD/mbdsamples_rab.txt) - using all three populations - HC/SS/NF  \n",
    "\n",
    "The predictor covariates - (4) - is the variable we will use to assess differential DNA methylation. Our data is both shell length and wet weight. \n",
    "- [size.macau.txt](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/data/size.macau.txt)\n",
    "\n",
    "NOTE: all 4 files need to have samples in the same order. They are all ordered using the mbd seq. sample numbers, 1-18. For the count data files, they automatically were ordered sequentially this way when I generated them. For the size covariate file, I re-orded samples in the [04-raw-count-files.Rmd notebook](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/code/04-raw-count-files.Rmd).  Katherine re-ordered her relatedness matrix in the [SOS_angsd.ipynb notebook](https://github.com/sr320/paper-oly-mbdbs-gen/blob/master/analyses/2bRAD/SOS_angsd.ipynb). \n",
    "\n",
    "Here are screen shots to confirm. \n",
    "\n",
    "_Hold for SSs of other dataframes from RStudio_\n",
    "\n",
    "Katherine's relatedness matrix order \n",
    "<img src=\"attachment:image.png\" width=\"600\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See if we can access MACAU "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      " MACAU version 1.00alpha, released on 06/05/2015\r\n",
      " implemented by Xiang Zhou\r\n",
      "\r\n",
      " type ./macau -h [num] for detailed helps\r\n",
      " options: \r\n",
      " 1: quick guide\r\n",
      " 2: file I/O related\r\n",
      " 3: fit binomial mixed models\r\n",
      " 4: note\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! /home/shared/macau/macau -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " FILE I/O RELATED OPTIONS\r\n",
      " -g        [filename]      specify input read count file name\r\n",
      "          format: site id, individual ID#1, individual ID#2, individual ID#3, ...\r\n",
      "                  siteID#1, expression for individual 1, expression for individual 2, expression for individual 3, ...\r\n",
      "                  siteID#2, expression for individual 1, expression for individual 2, expression for individual 3,  ...\r\n",
      "                  ...\r\n",
      "          missing expression value: NA\r\n",
      " -t        [filename]      specify input total read count file (optional); in the same format as above, or contain only one data row\r\n",
      " -p        [filename]      specify input: predictor variable file name\r\n",
      "          format: variable for individual 1\r\n",
      "                  variable for individual 2\r\n",
      "                  ...\r\n",
      "          missing predictor variable value: NA\r\n",
      " -n        [filename]      specify column number for predictor variable file (optional)\r\n",
      " -k        [filename]      specify input kinship/relatedness matrix file name\r\n",
      " -v        [filename]      specify input biological variation matrix file name (optional)\r\n",
      " -c        [filename]      specify input: covariates file name (optional)\r\n",
      "          format: covariate 1 for individual 1, ... , covariate c for individual 1\r\n",
      "                  covariate 1 for individual 2, ... , covariate c for individual 2\r\n",
      "                  ...\r\n",
      "          missing covariate value: NA\r\n",
      "          note: the intercept needs to be included\r\n",
      " -pace     [num]           specify terminal display update pace (default 1000 genes).\r\n",
      " -silence                  specify silence terminal output.\r\n",
      " -o        [prefix]        specify output file prefix (default \"result\")\r\n",
      "          output: prefix.assoc.txt and prefix.log.txt for analysis\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! /home/shared/macau/macau -h 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'/home/srlab/GitHub/paper-oly-mbdbs-gen/code'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# create a macau/ directory  \n",
    "\n",
    "! mkdir ../analyses/macau/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/srlab/GitHub/paper-oly-mbdbs-gen/analyses/macau\n"
     ]
    }
   ],
   "source": [
    "# MACAU writes output files to current directory, so move there \n",
    "%cd ../analyses/macau/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'/home/srlab/GitHub/paper-oly-mbdbs-gen/analyses/macau'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "siteID\tnumCs1\tnumCs2\tnumCs3\tnumCs4\tnumCs5\tnumCs6\tnumCs7\tnumCs8\tnumCs9\tnumCs10\tnumCs11\tnumCs12\tnumCs13\tnumCs14\tnumCs15\tnumCs16\tnumCs17\tnumCs18\r\n",
      "Contig0_39226\t11\t17\t10\t14\t9\t13\t9\t10\t18\t9\t4\t15\t10\t5\t8\t8\t8\t8\r\n",
      "Contig0_39234\t10\t7\t10\t5\t12\t11\t7\t8\t8\t7\t1\t10\t8\t5\t12\t5\t12\t10\r\n",
      "Contig0_71523\t16\t12\t13\t16\t11\t8\t11\t10\t15\t19\t17\t13\t24\t18\t28\t18\t17\t18\r\n",
      "Contig0_71533\t21\t17\t17\t17\t21\t13\t17\t20\t20\t27\t22\t17\t32\t23\t33\t22\t23\t22\r\n",
      "Contig0_71542\t21\t22\t16\t14\t15\t12\t28\t13\t18\t21\t19\t12\t35\t25\t32\t21\t16\t21\r\n",
      "Contig0_71546\t20\t19\t14\t12\t12\t15\t24\t13\t15\t19\t15\t11\t30\t21\t26\t20\t15\t16\r\n",
      "Contig0_71558\t20\t18\t12\t15\t13\t5\t25\t11\t18\t28\t21\t19\t29\t25\t29\t25\t23\t18\r\n",
      "Contig0_71617\t18\t13\t21\t17\t14\t8\t20\t19\t12\t20\t18\t12\t23\t24\t29\t17\t23\t17\r\n",
      "Contig0_71647\t19\t19\t14\t21\t15\t9\t21\t28\t19\t19\t26\t15\t20\t27\t33\t29\t28\t20\r\n"
     ]
    }
   ],
   "source": [
    "# Check out methylated count file\n",
    "! head ../../analyses/counts-filtered-meth.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "siteID\tcoverage1\tcoverage2\tcoverage3\tcoverage4\tcoverage5\tcoverage6\tcoverage7\tcoverage8\tcoverage9\tcoverage10\tcoverage11\tcoverage12\tcoverage13\tcoverage14\tcoverage15\tcoverage16\tcoverage17\tcoverage18\r\n",
      "Contig0_39226\t21\t22\t22\t23\t23\t21\t20\t16\t31\t17\t5\t23\t17\t10\t20\t13\t19\t24\r\n",
      "Contig0_39234\t24\t27\t27\t26\t28\t25\t23\t24\t36\t19\t7\t29\t19\t12\t25\t17\t23\t32\r\n",
      "Contig0_71523\t16\t12\t13\t18\t13\t8\t16\t10\t16\t23\t18\t14\t28\t21\t33\t21\t20\t18\r\n",
      "Contig0_71533\t21\t17\t18\t17\t21\t13\t23\t20\t21\t27\t22\t21\t33\t24\t36\t23\t23\t23\r\n",
      "Contig0_71542\t23\t23\t16\t14\t15\t12\t28\t14\t18\t21\t19\t12\t35\t26\t33\t23\t17\t22\r\n",
      "Contig0_71546\t20\t19\t15\t12\t14\t15\t26\t14\t16\t21\t15\t11\t31\t22\t30\t21\t15\t17\r\n",
      "Contig0_71558\t20\t19\t12\t16\t13\t5\t27\t11\t19\t29\t22\t19\t30\t30\t30\t25\t23\t19\r\n",
      "Contig0_71617\t22\t20\t23\t20\t19\t10\t22\t22\t25\t29\t21\t17\t28\t29\t36\t28\t25\t19\r\n",
      "Contig0_71647\t22\t19\t17\t23\t16\t13\t22\t33\t22\t20\t30\t18\t23\t29\t35\t35\t30\t25\r\n"
     ]
    }
   ],
   "source": [
    "# Check out total count file\n",
    "! head ../../analyses/counts-filtered-total.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.2\t17.41\r\n",
      "1.9\t20.43\r\n",
      "2.2\t25.33\r\n",
      "1.1\t19.38\r\n",
      "2.2\t26.79\r\n",
      "1.2\t19.8\r\n",
      "2.1\t20.54\r\n",
      "1.9\t19.5\r\n",
      "1.4\t18.43\r\n",
      "2.2\t21.02\r\n"
     ]
    }
   ],
   "source": [
    "# Confirm which column I want to use as predictor variable. \n",
    "#1= covariate file intercept \n",
    "#2 - wet weight (grams, in shell) <-- use this as covariate\n",
    "#3 - shell length (mm) <-- use this as predictor \n",
    "\n",
    "! head ../../data/predictors.size.macau.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t2.2\r\n",
      "1\t1.9\r\n",
      "1\t2.2\r\n",
      "1\t1.1\r\n",
      "1\t2.2\r\n",
      "1\t1.2\r\n",
      "1\t2.1\r\n",
      "1\t1.9\r\n",
      "1\t1.4\r\n",
      "1\t2.2\r\n"
     ]
    }
   ],
   "source": [
    "! head ../../data/cov.weight.macau.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\t0.001054\t0.024934\t0.010619\t0.005431\t0.469934\t1e-06\t0.096852\t0.018074\t7e-06\t1e-06\t0\t1e-06\t1e-06\t0\t0\t0\t1e-06\r\n",
      "0.001054\t0\t0.132783\t0.281738\t0.178604\t0.023215\t0.303803\t5.1e-05\t0.045462\t0\t1e-06\t2.8e-05\t3e-06\t0\t0\t0\t0.004357\t0\r\n",
      "0.024934\t0.132783\t0\t0.091518\t0.105258\t0.010814\t0.202782\t0.215806\t8.6e-05\t3.5e-05\t9e-06\t0.019306\t1.3e-05\t2e-06\t5e-06\t2.2e-05\t0.032574\t6e-06\r\n",
      "0.010619\t0.281738\t0.091518\t0\t0.178816\t0.028384\t0.269209\t0.003049\t0.01418\t0\t0\t3e-06\t0\t0\t0\t0\t1e-06\t0\r\n",
      "0.005431\t0.178604\t0.105258\t0.178816\t0\t1e-06\t0.505432\t8e-06\t0.052672\t1e-06\t3e-06\t0\t0\t0\t0\t0\t0\t0\r\n",
      "0.469934\t0.023215\t0.010814\t0.028384\t1e-06\t0\t0\t0.098489\t0.017889\t0\t0\t2e-06\t0\t1e-06\t1e-06\t0\t0\t0\r\n",
      "1e-06\t0.303803\t0.202782\t0.269209\t0.505432\t0\t0\t4.1e-05\t0.019627\t1e-06\t0\t2e-06\t0\t0\t0\t0\t1e-06\t0\r\n",
      "0.096852\t5.1e-05\t0.215806\t0.003049\t8e-06\t0.098489\t4.1e-05\t0\t0.000844\t3e-06\t3e-06\t1.5e-05\t0.00168\t0\t0.059391\t6e-06\t0.000188\t5e-06\r\n",
      "0.018074\t0.045462\t8.6e-05\t0.01418\t0.052672\t0.017889\t0.019627\t0.000844\t0\t0\t0\t0.000103\t1e-06\t0\t0\t0\t3e-06\t5e-06\r\n",
      "7e-06\t0\t3.5e-05\t0\t1e-06\t0\t1e-06\t3e-06\t0\t0\t0.342652\t0.251287\t0.286715\t0.231231\t0\t0.235662\t0.012039\t0.188723\r\n"
     ]
    }
   ],
   "source": [
    "! head ../2bRAD/HCSSmbdsamples_rab.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the following options: \n",
    "    \n",
    "`-g`  specify the methylated read counts file name.  \n",
    "`-t`  specify the total read counts file name.  \n",
    "`-p`  specify the predictor variable file name.  \n",
    "`-n 2` specify which predictor variable column to use in analysis. In our case column 2 (length in mm)  \n",
    "`-k`  specify the kinship/relatedness matrix file name.  \n",
    "`-bmm`  specifies binomial mixed model.  \n",
    "`-o`  specify output file prefix (default “result”).  \n",
    "\n",
    "Run started at , finished at . "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Files ... \n",
      "## number of total individuals = 18\n",
      "## number of analyzed individuals = 18\n",
      "## number of covariates = 2\n",
      "## number of total genes/sites = 19648\n",
      "Performing Analysis ==================================================100.00%\n"
     ]
    }
   ],
   "source": [
    "! /home/shared/macau/macau \\\n",
    "-g ../counts-filtered-meth.txt \\\n",
    "-t ../counts-filtered-total.txt \\\n",
    "-p ../../data/predictors.size.macau.txt \\\n",
    "-n 2 \\\n",
    "-c ../../data/cov.weight.macau.txt \\\n",
    "-k ../2bRAD/HCSSmbdsamples_rab.txt \\\n",
    "-bmm \\\n",
    "-o 20200205-macau"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20190812-heatmap-p001.pdf  macau-all-loci.10x75.bed\r\n",
      "allLoci.geneinfo.txt       macau-all-loci.5x75.bed\r\n",
      "macau.10x75.c.tab          macau-all-loci.bed\r\n",
      "macau-10x75perc.bed        macau-any10x.bed\r\n",
      "macau.10xall.annot.csv     MACAU.geneinfo.txt\r\n",
      "macau.20190812.c.tab       macau-heatmap.10x75.cluster.pdf\r\n",
      "macau.5x75.c.tab           macau-heatmap.10x75.length.pdf\r\n",
      "macau75.features.csv       macau-heatmap.10x75.tree.pdf\r\n",
      "macau75.GO.csv             macau.sign.perc.meth.10x75.bed\r\n",
      "macau75.GO.txt             macau.sign.perc.meth.5x75.bed\r\n",
      "macau75.sig.csv            \u001b[0m\u001b[01;34moutput\u001b[0m/\r\n",
      "macau-all10x.bed\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id\tn\tacpt_rate\tbeta\tse_beta\tpvalue\th\tse_h\tsigma2\tse_sigma2\talpha0\tse_alpha0\talpha1\tse_alpha1\r\n",
      "Contig0_39226\t18\t4.795e-01\t-1.128e-02\t3.639e-02\t7.565e-01\t6.827e-01\t2.488e-01\t1.236e+00\t3.107e+00\t8.535e-01\t6.794e-01\t-1.886e-01\t1.683e-01\r\n",
      "Contig0_39234\t18\t3.973e-01\t3.811e-02\t3.652e-02\t2.968e-01\t7.570e-01\t2.258e-01\t1.503e+00\t3.325e+00\t-1.283e+00\t7.052e-01\t-1.099e-01\t1.626e-01\r\n",
      "Contig0_71523\t18\t2.984e-01\t-9.373e-02\t7.450e-02\t2.083e-01\t7.189e-01\t2.704e-01\t8.320e+00\t1.331e+01\t3.175e+00\t1.366e+00\t6.219e-01\t3.489e-01\r\n",
      "Contig0_71533\t18\t3.405e-01\t1.088e-02\t1.502e-01\t9.423e-01\t3.152e-01\t1.767e-01\t6.650e+00\t5.669e+00\t3.392e+00\t2.551e+00\t9.812e-02\t6.823e-01\r\n",
      "Contig0_71542\t18\t3.138e-01\t-7.261e-03\t9.611e-02\t9.398e-01\t6.044e-01\t2.586e-01\t4.163e+00\t6.901e+00\t4.014e+00\t1.740e+00\t1.720e-02\t5.362e-01\r\n",
      "Contig0_71546\t18\t3.150e-01\t-8.721e-02\t8.660e-02\t3.139e-01\t6.522e-01\t2.553e-01\t4.602e+00\t9.124e+00\t4.930e+00\t1.690e+00\t2.095e-01\t3.717e-01\r\n",
      "Contig0_71558\t18\t2.732e-01\t-6.461e-02\t1.233e-01\t6.002e-01\t5.574e-01\t2.850e-01\t1.038e+01\t1.787e+01\t4.911e+00\t2.233e+00\t1.548e-01\t6.232e-01\r\n",
      "Contig0_71617\t18\t3.085e-01\t-1.921e-02\t5.806e-02\t7.408e-01\t6.385e-01\t2.760e-01\t4.597e+00\t9.327e+00\t9.871e-01\t1.067e+00\t3.325e-01\t2.662e-01\r\n",
      "Contig0_71647\t18\t3.201e-01\t2.459e-02\t4.587e-02\t5.918e-01\t7.808e-01\t1.955e-01\t2.808e+00\t4.401e+00\t1.821e+00\t9.188e-01\t-7.683e-02\t2.048e-01\r\n"
     ]
    }
   ],
   "source": [
    "# Peek at the results file \n",
    "! head output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id\tn\tacpt_rate\tbeta\tse_beta\tpvalue\th\tse_h\tsigma2\tse_sigma2\talpha0\tse_alpha0\talpha1\tse_alpha1\r\n"
     ]
    }
   ],
   "source": [
    "# List column names - p-value is 6th column (use for indexing)\n",
    "! head -n 1 output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1089\r\n"
     ]
    }
   ],
   "source": [
    "# Count # hits with p-value <0.05\n",
    "! awk '$6<0.05{pvalue++} END{print pvalue+0}' \\\n",
    "output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "292\r\n"
     ]
    }
   ],
   "source": [
    "# Count # hits with p-value <0.01\n",
    "! awk '$6<0.01{pvalue++} END{print pvalue+0}' \\\n",
    "output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "65\r\n"
     ]
    }
   ],
   "source": [
    "# Count # hits with p-value <0.001\n",
    "! awk '$6<0.001{pvalue++} END{print pvalue+0}' \\\n",
    "output/20200205-macau.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##\r\n",
      "## MACAU Version = 1.00alpha\r\n",
      "##\r\n",
      "## Command Line Input = /home/shared/macau/macau -g ../counts-filtered-meth.txt -t ../counts-filtered-total.txt -p ../../data/predictors.size.macau.txt -n 2 -c ../../data/cov.weight.macau.txt -k ../2bRAD/HCSSmbdsamples_rab.txt -bmm -o 20200205-macau \r\n",
      "##\r\n",
      "## Date = Fri Feb  7 15:31:45 2020\r\n",
      "##\r\n",
      "## Summary Statistics:\r\n",
      "## number of total individuals = 18\r\n",
      "## number of analyzed individuals = 18\r\n",
      "## number of total genes/sites = 19648\r\n",
      "## number of analyzed genes/sites = 39296\r\n",
      "## number of covariates = 2\r\n",
      "##\r\n",
      "## Computation time:\r\n",
      "## total computation time = 20.1332 min \r\n",
      "## computation time break down: \r\n",
      "##      time on eigen-decomposition = 0.5475 min \r\n",
      "##      time on calculating matrix-vector multiplication = 1.64117 min \r\n",
      "##      time on sampling Z = 1.464 min \r\n",
      "##      time on sampling MW = 1.7635 min \r\n",
      "##      time on sampling Hyperparameters = 13.919 min \r\n",
      "##      time on sampling BAU = 1.66033 min \r\n",
      "##\r\n"
     ]
    }
   ],
   "source": [
    "# Review run log \n",
    "! cat output/20200205-macau.log.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
